---
title: "Annotating Rau-Patterson snRNAseq"
author: "Brian Gural"
date: "2023-09-13"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 10, message = F, warning = F, error = T)
```

```{r load libs}
libs <- c("tidyverse", "Seurat", "SeuratDisk", "SCpubr", "Biobase", "MuSiC", 
          "reshape2", "SingleCellExperiment") # list libraries here
lapply(libs, require, character.only = T)
rm(libs)
```

```{r load data}

sn <- LoadH5Seurat("/proj/raulab/users/brian/r_projects/cc_composition/data/processed/single_cell/merges/rau_patterson/09132023/doublets_below_3.h5seurat")

#Load Lusis TPM datasets
bulk_cont <- read.csv("lusis/data/raw/HFpEF_wholeheart_TPMs.txt", row.names = 1) 
bulk_iso <- read.csv("lusis/data/raw/ISO_ventricles_TPMs.txt", row.names = 1) 

# Make phenotypes 
phenos <- data.frame(Sub = c(colnames(bulk_cont[-c(1:6)]),colnames(bulk_iso[-c(1:6)]))) |> 
          mutate(treatment = factor(case_when(
            str_detect(Sub, "R1") ~ "Control",
            str_detect(Sub, "GF") ~ "Treatment")))

```

```{r merge isoforms , fig.width = 14, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
bulk_clean <- bulk_cont |> select(-c("Gene.stable.ID","Exon.stable.ID","Gene.start..bp.","Gene.end..bp.", "Gene.type")) |> 
  group_by(Gene.name) |> 
  summarize_all(sum) |> 
  column_to_rownames("Gene.name")

bulk_clean_iso <- bulk_iso |> select(-c("Gene.stable.ID","Exon.stable.ID","Gene.start..bp.","Gene.end..bp.", "Gene.type")) |> 
  group_by(Gene.name) |> 
  summarize_all(sum) |> 
  column_to_rownames("Gene.name")

common.genes <- row.names(bulk_clean)[row.names(bulk_clean) %in% row.names(bulk_clean_iso)]

bulk.all <- cbind(bulk_clean[common.genes,], bulk_clean_iso[common.genes,])

rm(bulk_clean)
rm(bulk_clean_iso)
gc()
```

```{r scran find markers, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}

# Update sn with annotations
sn.mark <- sn |>
      #subset(idents = c(0,1,2,3,4,5)) |> 
      subset(features = rownames(sn)[rownames(sn) %in% rownames(bulk.all)])

sn.sce <- sn.mark |> 
  as.SingleCellExperiment(assay = "RNA") |>
  as("SummarizedExperiment")

# Add gene names
rownames(sn.sce) <- rownames(sn.mark@assays$RNA@counts)

# Get markers based on expression ratios between clusters and 1vAll comparisons

scn.markers  <- scran::findMarkers(sn.sce, groups = Idents(sn.mark), pval.type = "some", assay.type = "counts")


.getMarkers <- function(type){
marker <- scn.markers@listData[[type]] |>
  as.data.frame() %>%
  dplyr::filter(summary.logFC >= 1.1 |
                   summary.logFC <= -1.1) |>
  arrange(FDR) |>
  slice_head(n = 15) |>
  dplyr::select(p.value, FDR, summary.logFC) |>
  mutate(celltype = type) |> 
  rownames_to_column(var = "gene")
return(marker)}

# Collapse list of data frames
all.markers <- lapply(levels(Idents(sn.mark)), function(x){.getMarkers(x)}) |>
               purrr::reduce(full_join)

all.markers |>
  group_by(celltype) |>
  arrange(desc(summary.logFC)) |>
  slice_head(n = 3) 

gc()
```

```{r print gene lists}

# Genes here are printed and manuallly entered into ToppGene
# https://toppgene.cchmc.org/
genes <- function(x){ all.markers |> 
                      subset(celltype == x) |> 
                      pull(gene) |> 
                      as.data.frame() |> 
                      print(row.names = F, print.keys = F)
}

cell.types <- c("Endothelial Cells",
                "Cardiomyocytes",
                "Fibroblast",
                #"VSMC/Pericytes",
                "Monocytes/MF")
                #"SMC"
                #"Cardiac Neuron",
                #"?Matrix FB, Lung?")
                #"unclear",#"??EC, CA??"
                #"unclear")#"???")

sn.mark <- sn |>
 #subset(idents = seq(0,length(cell.types)-1,1)) 
 subset(idents = c(0,1,2,4))
names(cell.types) <- levels(sn.mark)
sn.mark <- RenameIdents(sn.mark, cell.types)
```

```{r music}

bulk.es.exp <- bulk.all[all.markers$gene,] |> 
  sapply(as.integer) %>%
  ExpressionSet(assayData = .) |> 
  exprs()
row.names(bulk.es.exp) <- all.markers$gene


# Run MuSiC with subset genes

# Convert to SingleCellExperiment
sce <- as.SingleCellExperiment(sn.mark, assay = "RNA")
  
# Exclude specified clusters
cells <- levels(Idents(sn.mark))
cells <- cells[!(cells %in% c("unclear")) & !is.na(cells)]
  
# Use MuSiC to estimate cell type proportions
decon <- music_prop(bulk.mtx = bulk.es.exp, sc.sce = sce, markers = all.markers$gene,
                      clusters = "ident", samples = "orig.ident",
                      select.ct = cells)
  
# Turn MuSiC output into graph-friendly dataframe
decon.melt <- reshape2::melt(decon$Est.prop.weighted)
colnames(decon.melt) = c('Sub', 'CellType', 'Prop')

rm(decon)
rm(sce)
rm(cells)

```

```{r reformat data}
# Divide fraction and whole data

decon.melt <- decon.melt|> 
                mutate(Treatment = factor(case_when(
            str_detect(Sub, "R1") ~ "Control",
            str_detect(Sub, "GF") ~ "Treatment")))

# Unmelt for clustering 
decon.wide <- decon.melt  |> 
              select(Sub, CellType, Prop) |> 
              pivot_wider(names_from = "Sub", values_from = "Prop") |> 
              column_to_rownames("CellType") %>%
              mutate_all(as.numeric)

# Reorder the factor levels of CellType so that they are ordered by sum of their values across groups
decon.melt$CellType <- reorder(decon.melt$CellType, -decon.melt$Prop, FUN = sum)

```


```{r comp plots, echo=FALSE, fig.width=15, fig.height = 8}
library(RColorBrewer)
brewer.pal(n=8,"Paired")

my_palette <- c("#A6CEE3", 
                #"#1F78B4", 
                "#FDBF6F"#, 
                #"#FF7F00"
                )
legend.names <- c("Sham_1","Sham_2", "TAC_1", "TAC_2")


ordered_df <- decon.melt %>%
  filter(CellType == "Cardiomyocytes") %>%  # Filter rows where CellType is "Cardiomyocytes"
  arrange(Prop) %>%  # Sort these rows by Prop in descending order
  pull(Sub)  # Extract the Sub values in this order
# Use the ordered Sub values to reorder the original data frame
decon.melt <- decon.melt %>%
  mutate(Sub = factor(Sub, levels = ordered_df)) %>%
  arrange(Sub)
decon.melt$CellType_wrap = str_wrap(decon.melt$CellType, width = 12)


cell.type.order <-  decon.melt %>%
  group_by(CellType_wrap) |> 
  mutate(mean = mean(Prop)) |> 
  arrange(desc(mean)) |>
  pull(CellType_wrap) |> 
  unique()

design <- "AA
           BB"

# Generate boxplot
comp_celltype <- decon.melt   %>%
    ggplot(aes(x = factor(CellType_wrap, levels = as.character(cell.type.order)), y = Prop, fill = Treatment)) +
   #geom_boxplot(position = position_dodge(0.9), width = 0.9, color = "black") +
   geom_bar(stat = "summary",position = position_dodge(0.95),  fun = mean,width = 0.9,  color = "black", alpha = 0.8) +
    geom_jitter(inherit.aes = T, 
                position = position_jitterdodge(jitter.width = 0.05, dodge.width = 0.9),
                size = 4, alpha = 0.5) +
    theme(axis.text.x = element_text(color = "black", size = 28, angle =25, vjust = 0.5),
          axis.text.y = element_text(color = "black", size = 28),
          axis.ticks = element_blank(),
          legend.position = c(0.9, 1),
          legend.justification = c("right", "top"),
          legend.box.just = "right",
          legend.margin = margin(6, 6, 6, 6),
          strip.text = element_text(size = 20),
          title = element_text(size = 30),
          legend.text = element_text(size = 32),
          panel.background = element_rect(fill='transparent'),
          plot.background = element_rect(fill='transparent', color=NA),
          #panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          legend.background = element_rect(fill='transparent'),
          legend.box.background = element_rect(fill='transparent'),
          axis.title.x = element_blank(),
          plot.margin = unit(c(1,1,1,1), units = "cm"),
          axis.title.y = element_text(size = 30)) +
    labs(y = "Proportion", 
         fill = "Treatment") +
    scale_fill_manual(values = my_palette)
  
comp_celltype

```