---
title: "Jensen decon with many reference datasets"
author: "Brian Gural"
date: "6/6/2023"
output:
  html_document:
    toc: true
    toc_float: true
---




```{r setup, include=FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Introduction and Context\
The aim of this analysis is to estimate the cellular composition of bulk-RNAseq reads from mouse left ventricular tissue. We'll find cell-type-specific gene expression reference profiles with snRNAseq. These two single nucleus samples are aggregates of several adult, male B6 from Christoph's post-doc.\
\
This document explains the workflow in the order in which the data processing occurred. You can jump to portions with the table of contents on the left.\

## Load data and packages\


```{r load libs, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# load libraries
libs <- c("Seurat", "ggplot2", "DESeq2", "patchwork","SeuratDisk", "MuSiC", "reshape2",
          "tidyverse", "SingleCellExperiment","harmony", "SCpubr","shiny", 
          "AUCell", "viridis", "gplots", "scales", "ggrepel", "gridExtra", "scCustomize",
          "httr","readxl","matrixStats", "TabulaMurisSenisData") # list libraries here
lapply(libs, require, character.only = T)
source("jensen/scripts/functions/decon_all.R")
```

```{r load data, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# bulk
bulk <- read.csv("jensen/data/processed/jensen_rau_froese_cpm", row.names = 1)
bulk_pheno <- read.csv("jensen/data/processed/jensen_rau_froese_pheno", row.names = 1)

# Single cell
sn.rau <- LoadH5Seurat("jensen/data/raw/b6_sn.h5seurat")
sn.tm  <- LoadH5Seurat("jensen/data/processed/tabula_muris.h5seurat")
sn.tm$orig.ident <- paste0("tm_", sn.tm$mouse.id)
sn.tm <- subset(sn.tm, tissue_free_annotation == "Heart")

sn.wu  <- LoadH5Seurat("jensen/data/processed/single_cell/wu_2021.h5seurat")
sn.mt  <- LoadH5Seurat("jensen/data/processed/single_cell/martini_2019.h5seurat")

sn.rau$origin <- "rau"
sn.mt$origin <- "martini"

sn.all <- merge(sn.rau, sn.tm) |>
            merge(sn.wu) |>
            merge(sn.mt)

sn.all$PercentMito <- PercentageFeatureSet(sn.all, pattern = "^mt-")
```

```{r cluster 1, message=FALSE, warning=FALSE, cache=FALSE, include=F}
sn.clust.1 <- ClusterSeurat(sn.all, max.rna.ft = 10000,
                            max.mt.pt = 5, res = 0.03, regress.by = c("origin"))
```

```{r cluster1.dim, echo=FALSE, fig.height=14, fig.width=10, message=FALSE, warning=FALSE, cache=T}
# standard visuals for clusters and mitocondrial contaminants
dim.ft  <- c("seurat_clusters", "origin", "orig.ident")
feat.ft <- c("PercentMito", "nCount_RNA")
vln.ft <- c("PercentMito")
vln.groups <- c("seurat_clusters")

plotUMAP(data = sn.clust.1,
         dim.ft = dim.ft,
         feat.ft = feat.ft,
         nrow = 3, 
         ncol = 2)

```

```{r load broad markers, echo=F, fig.height=10, fig.width=8, message=FALSE, warning=FALSE, cache = T, cache=T}
# load markers
sn.clust.2 <- sn.clust.1
markers.broad <- read.csv("jensen/data/processed/external/mclellan_2020/mclellan_cell_markers_broad.csv")
```

\
```{r broad.cell.types AUC, echo=FALSE, fig.height=4, fig.width=7, message=FALSE, warning=FALSE, cache=TRUE}
# Split 'markers' into separate data frames for each unique subcluster
subclusters <- split(markers.broad, markers.broad$cluster)
# Extract the gene names from each subcluster data frame
subcluster_genes <- lapply(subclusters, function(x) x$gene)
# Create a list with named elements corresponding to each subcluster
geneSets <- setNames(subcluster_genes, names(subclusters))
# make expression matrix of single cell
my.expr <-  sn.clust.2@assays$RNA@counts
# Find  AUC for each cell by each type
cells_AUC <- AUCell_run(my.expr, geneSets)
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=F, assign=TRUE)

# get the thresholds and report when/how they're passdd
selectedThresholds <- getThresholdSelected(cells_assignment)

auc.val <- cells_AUC@assays@data@listData$AUC |>
              as.data.frame() |>
              t()

# Apply process_cell_id function to each row of auc.val using lapply()
broad_cell_types_list <- lapply(1:nrow(auc.val), function(i) processCellId(auc.val[i, ]))

# Convert the list to a data frame
broad.cell.types <- data.frame(BroadCellType = unlist(broad_cell_types_list), stringsAsFactors = FALSE)
rownames(broad.cell.types) <- rownames(auc.val)

sn.clust.2 <- AddMetaData(sn.clust.2, broad.cell.types)
```

```{r fine.cell.types, echo=FALSE, fig.height=4, fig.width=7, message=FALSE, warning=FALSE, cache=T}
# load markers
markers.fine <- read.csv("jensen/data/processed/external/mclellan_2020/mclellen_cell_markers.csv")

# Split 'markers' into separate data frames for each unique subcluster
subclusters <- split(markers.fine, markers.fine$subcluster)
# Extract the gene names from each subcluster data frame
subcluster_genes <- lapply(subclusters, function(x) x$gene)
# Create a list with named elements corresponding to each subcluster
geneSets <- setNames(subcluster_genes, names(subclusters))
# make expression matrix of single cell
my.expr <-  sn.clust.2@assays$RNA@counts
# Find  AUC for each cell by each type
cells_AUC <- AUCell_run(my.expr, geneSets)
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=F, assign=TRUE)

# get the thresholds and report when/how they're passdd
selectedThresholds <- getThresholdSelected(cells_assignment)

auc.val <- cells_AUC@assays@data@listData$AUC |>
              as.data.frame() |>
              t()

# Apply process_cell_id function to each row of auc.val using lapply()
fine_cell_types_list <- lapply(1:nrow(auc.val), function(i) processCellId(auc.val[i, ]))

# Convert the list to a data frame
fine.cell.types <- data.frame(FineCellType = unlist(fine_cell_types_list), stringsAsFactors = FALSE)
rownames(fine.cell.types) <- rownames(auc.val)

sn.clust.2 <- AddMetaData(sn.clust.2, fine.cell.types)
```

### Label Clusters\

```{r dim broad/fine, echo=FALSE, fig.height=6, fig.width=10, message=FALSE, warning=FALSE,}
# standard visuals for clusters and mitocondrial contaminants
dim.ft  <- c("BroadCellType", "FineCellType")
plotUMAP(data = sn.clust.1,
         dim.ft = dim.ft,
         feat.ft = NULL,
         nrow = 1, 
         ncol = 2)

```
\
And these are the clusters we found earlier, for reference:\
\
```{r dim broad/fine ref, echo=FALSE, fig.height=10, fig.width=12, message=FALSE, warning=FALSE,fig.align = 'center'}
# standard visuals for clusters and mitocondrial contaminants
dim.ft  <- c("seurat_clusters")
feat.ft <- "PercentMito"
plotUMAP(data = sn.clust.2,
         dim.ft = dim.ft,
         feat.ft = feat.ft,
         nrow = 1, 
         ncol = 2)

```
\

### Defining Seurat clusters\


```{r heatmap for all celltypes, echo=FALSE, fig.height=12, fig.width=10, message=FALSE, warning=FALSE, eval=T}
# Create a contingency table
contingency_table <- table(c(sn.clust.2$BroadCellType,sn.clust.2$FineCellType),
                           c(sn.clust.2$seurat_clusters,sn.clust.2$seurat_clusters))

# Perform the chi-squared test
chi_squared_test <- chisq.test(contingency_table)

# Calculate the standardized residuals
observed_frequencies <- contingency_table
expected_frequencies <- chi_squared_test$expected
standardized_residuals <- (observed_frequencies - expected_frequencies) / sqrt(expected_frequencies)


col_palette <- viridis(25)
# Create a heatmap of the standardized residuals
heatmap.2(standardized_residuals, 
          trace = "none", 
          col = col_palette,
          main = "Residuals, All Cell Types",
          xlab = "Seurat Clusters",
          ylab = "All Cell Types",
          margins = c(8, 12),
          key.title = "Standardized Residuals",
          srtCol = 0,
          cexRow = 1.2)

```
\

```{r label clusters w chisq for celltypes, echo=FALSE, fig.height=6, fig.width=10, message=FALSE, warning=FALSE}
# Find the index of the largest value in each column of the normalized_contingency_table
max.indices <- apply(standardized_residuals, 2, which.max)
# Get the corresponding BroadCellType for each Seurat cluster
assigned.cell.types <- rownames(standardized_residuals)[max.indices] |>
                          make.unique()

#rename the clusters
names(assigned.cell.types) <- levels(sn.clust.2)
sn.clust.2 <- RenameIdents(sn.clust.2, assigned.cell.types)
SaveH5Seurat(sn.clust.2, "jensen/data/processed/single_cell/06062023")

```

Here's what our final clusters and labels look like: \
```{r final cluster labels, echo=FALSE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE}
do_DimPlot(sn.clust.2, reduction = "umap", label = T, repel = T) + NoLegend()
```
```{r final cluster labels, echo=FALSE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE}
do_DimPlot(sn.clust.2, reduction = "umap",group.by = "origin", label = T, repel = T, shuffle = T) + NoLegend()
```

## MuSiC and composition estimates\


```{r cluster2.dim, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# music_prop needs exprs matrix
sn.sce <- as.SingleCellExperiment(sn.clust.2, assay = "RNA")
bulk.es <- ExpressionSet(assayData = as.matrix(bulk))
bulk.es <- exprs(bulk.es)
# estimate bulk composition with music
# Find the index of the specified cell type
labels <- standardized_residuals |>
  as.data.frame() |>
  group_by(Var2) |>
  summarise(Value = max(Freq))

good_labels <- labels[which(labels$Value > 40), 1]
cell_types <-levels(Idents(sn.clust.2))[good_labels$Var2]
# Access the cell type using the index

decon <- music_prop(bulk.mtx = 2^bulk.es, sc.sce = sn.sce, markers = NULL,
                    clusters = "ident", samples = "orig.ident",
                    select.ct = cell_types)
```


```{r cluster2.feat , fig.width = 7, fig.height= 4, echo=FALSE, message=FALSE, warning=FALSE}
# turn music output into graph-friendly dataframe
# also adds ratios of nuclei clusters from subject-matched samples
decon.melt = melt(decon$Est.prop.weighted)
colnames(decon.melt) = c('Sub', 'CellType', 'Prop')
decon.melt$CellType = factor(decon.melt$CellType, levels = unique(decon.melt$CellType))

decon.melt$CellType = factor(decon.melt$CellType, levels = unique(decon.melt$CellType))
decon.melt$Prop <- as.numeric(decon.melt$Prop)

decon.melt$genotype <- "wt"
decon.melt[grepl("A", decon.melt$Sub), 4] <- "ko"
decon.melt$treatment <- "control"
decon.melt[grepl("Lx", decon.melt$Sub), 5] <- "mi"
decon.melt$type <- "whole"
decon.melt[grepl("RNA", decon.melt$Sub), 6] <- "fraction"

decon.melt$origin <- "Jensen"
decon.melt[grepl("B6", decon.melt$Sub), 4] <- "Rau"
decon.melt[grepl("Sham", decon.melt$Sub), 4] <- "Froese"

```

```{r music plot controls, fig.width = 12, fig.height= 7, echo=FALSE, message=FALSE, warning=FALSE}
ggplot(decon.melt,
       aes(x=Sub, y=Prop, fill=CellType))  +
  geom_bar(stat='identity',
           position = "fill",
           width = 0.5,
           color = "black")+
  scale_fill_viridis_d(name = "Cell Type")+
  facet_wrap(~genotype, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
  ylab("Proportion") +
  theme( axis.text.x = element_text(color = "black", size = 8, angle = 30),
         strip.text = element_text(size = 20),
         title = element_text(size = 20),
         legend.text = element_text(size = 18),
         axis.ticks.x = element_blank()) +
  xlab("Samples (5% mitocondrial threshold in reference)")

```