---
title: "dataset_inclusion_test"
author: "Brian Gural"
date: "6/15/2023"
output: html_document
---

```{r setup, include=FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Introduction and Context\
The aim of this analysis is to estimate the cellular composition of bulk-RNAseq reads from mouse left ventricular tissue. We'll find cell-type-specific gene expression reference profiles with snRNAseq. These two single nucleus samples are aggregates of several adult, male B6 from Christoph's post-doc.\
\
This document explains the workflow in the order in which the data processing occurred. You can jump to portions with the table of contents on the left.\

## Load data and packages\


```{r load libs, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# load libraries
libs <- c("Seurat", "ggplot2", "DESeq2", "patchwork","SeuratDisk", "MuSiC", "reshape2",
          "tidyverse", "SingleCellExperiment","harmony", "SCpubr","shiny", 
          "AUCell", "viridis", "gplots", "scales", "ggrepel", "gridExtra", "scCustomize",
          "httr","readxl","matrixStats", "TabulaMurisSenisData") # list libraries here
lapply(libs, require, character.only = T)
source("jensen/scripts/functions/decon_all.R")
```

```{r load data, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# bulk
bulk <- read.csv("jensen/data/processed/jensen_rau_froese_cpm", row.names = 1)
bulk_pheno <- read.csv("jensen/data/processed/jensen_rau_froese_pheno", row.names = 1)

# Single cell
sn <- LoadH5Seurat("jensen/data/processed/single_cell/06152023.h5seurat")

```

```{r  make origin combinations, message=FALSE, warning=FALSE, cache=FALSE, include=F}


sources <- unique(sn$origin)[unique(sn$origin) != "rau"]
# Create an empty data frame to store the combinations
combinations_df <- data.frame(matrix(ncol = length(sources), nrow = 0))
# Loop over the number of sources in each combination
for (i in 1:length(sources)) {
  
  # Get the combinations
  combinations <- combn(sources, i, simplify = FALSE)
  
  # Add to the data frame, padding with NAs as necessary
  for (j in 1:length(combinations)) {
    # Pad with NAs
    combination <- combinations[[j]]
    combination <- c(combination, rep(NA, length(sources) - length(combination)))
    
    # Add to the data frame
    combinations_df <- rbind(combinations_df, combination, stringsAsFactors = FALSE)
  }
}

combinations_df <- t(combinations_df)
# make sure Rau appears in all
combinations_df <- rbind(combinations_df, 
                         rep("rau", length(combinations_df)))

# make a new seurat for each combination
# probably not a good idea

# assuming 'sn' is your Seurat object
sn.combos <- list()

for (i in 1:length(combinations_df)) {
  # extract sources from the combinations_df
  sources <- na.omit(combinations_df[,i])
  
  # subset the Seurat object based on the 'origin' column in the meta.data
  subset_sn <- subset(sn, origin %in% sources) 
  
  # add the new Seurat object to the list
  sn.combos[[paste(sources, collapse = "_")]] <- subset_sn
}
```


```{r ClusterSeurat, message=FALSE, warning=FALSE, cache=FALSE, include=F}
sn.combos <- lapply(sn.combos, function(seurat) {
  ClusterSeurat(seurat, max.rna.ft = 25000, max.mt.pt = 5, res = 0.03, regress.by = c("origin"))
})

```

```{r labels and clusters, echo=FALSE, fig.height=40, fig.width=12, message=FALSE, warning=FALSE, cache=T}
# standard visuals for clusters and mitocondrial contaminants
for(i in seq(1:length(sn.combos))){
sn.combos[[i]]$BroadCellType[sn.combos[[i]]$BroadCellType == "NA"] <- NA
} 


dim.plots <- vector("list", length(sn.combos)*2)

for(i in seq(1:length(sn.combos))*2){
    dim.plots[[i-1]] <- do_DimPlot(sn.combos[[i/2]], 
                             reduction = 'umap',
                             label = T,
                             repel = T,
                             shuffle = T) + 
                          NoLegend() + 
                          ggtitle(names(sn.combos)[[i/2]])
    dim.plots[[i]] <- do_DimPlot(sn.combos[[i/2]], 
                             reduction = 'umap', 
                             group.by = "BroadCellType",
                             label = T,
                             repel = T,
                             shuffle = T) + 
                          NoLegend() 

} 

patchwork::wrap_plots(dim.plots,
                          ncol = 2,
                          nrow = 7)


```

```{r functions, echo=FALSE, fig.height=12, fig.width=10, message=FALSE, warning=FALSE, eval=T}
AssignAndFilterClusters <- function(seurat, res.thresh = 80, ratio.thresh = 2, min.cell = 400) {
  # Create a contingency table
  contingency_table <- table(c(seurat$BroadCellType),
                             c(seurat$seurat_clusters), 
                             useNA = "ifany")

  # Perform the chi-squared test
  chi_squared_test <- chisq.test(contingency_table)

  # Calculate the standardized residuals
  observed_frequencies <- contingency_table
  expected_frequencies <- chi_squared_test$expected
  standardized_residuals <- (observed_frequencies - expected_frequencies) / sqrt(expected_frequencies)

  rownames(standardized_residuals)[is.na(rownames(standardized_residuals))] <- "unlabeled"

  # Exclude clusters below specific max residual threshold
  max <- apply(standardized_residuals, 2, max)
  max.exclude <- max < res.thresh

  # Exclude clusters whose max residual is less than double the unlabelled residual
  unlabel.exclude <- abs(max/standardized_residuals[nrow(standardized_residuals),]) < ratio.thresh

  # Exclude clusters with few cells 
  bad.clusts <- table(Idents(seurat))[table(Idents(seurat)) < min.cell]

  # Get final exclusion list
  all.exclude <- unlabel.exclude | unlabel.exclude
  
  # Find the index of the largest value in each column of the standardized_residuals
  max.indices <- apply(standardized_residuals, 2, which.max)

  max.indices[all.exclude] <- NA

  # Get the corresponding BroadCellType for each Seurat cluster
  assigned.cell.types <- rownames(standardized_residuals)[max.indices] 

  # Rename the clusters
  names(assigned.cell.types) <- levels(seurat)
  seurat <- RenameIdents(seurat, assigned.cell.types)

  # Remove very small clusters 
  bad.clusts <- names(table(Idents(seurat)))[table(Idents(seurat)) < min.cell]
  Idents(seurat)[which(Idents(seurat) %in% bad.clusts)] <- NA
  
  return(seurat)
}

EstimateCellTypeProportions <- function(seurat, bulk.es, for.aitchison = F, cells_exclude = c("unlabeled", "NA")) {
  
  # Convert to SingleCellExperiment
  seurat_sce <- as.SingleCellExperiment(seurat, assay = "RNA")

  # Exclude specified clusters
  cells <- levels(Idents(seurat))
  cells <- cells[!(cells %in% cells_exclude) & !is.na(cells)]
  
  # Use MuSiC to estimate cell type proportions
  decon <- music_prop(bulk.mtx = 2^bulk.es, sc.sce = seurat_sce, markers = NULL,
                      clusters = "ident", samples = "orig.ident",
                      select.ct = cells)
  if(for.aitchison == T){
  return(decon)
  }
  # Turn MuSiC output into graph-friendly dataframe
  decon.melt = reshape2::melt(decon$Est.prop.weighted)
  colnames(decon.melt) = c('Sub', 'CellType', 'Prop')
  decon.melt$combination <- paste(unique(seurat$origin), collapse = "_")
  return(decon.melt)
}

CreateSimFractions <- function(seurat, sim_samples, cell_dict, purity = 0.95) {
  
  # Get the cell types
  cell_types <- levels(seurat@active.ident)
  
  # Initialize the sim.fractions dataframe
  sim_fractions <- matrix(nrow = nrow(sim_samples), ncol = length(cell_types)) %>%
    as.data.frame()
  row.names(sim_fractions) <- sim_samples$id
  colnames(sim_fractions) <- cell_types
  
  # Fill in the dataframe
  for(i in rownames(sim_fractions)){
    name1 <- sim_samples[which(sim_samples$id == i), 2]
    major <- cell_dict[which(cell_dict$bulk.pheno == name1), 2]
    sim_fractions[i, major] <- purity
    sim_fractions[i, cell_types[which(cell_types != major)]] <- (1 - purity) / length(cell_types)
  }
  
  return(sim_fractions)
}

```
\

```{r label clusters w chisq for celltypes, echo=FALSE, fig.height=6, fig.width=10, message=FALSE, warning=FALSE}


sn.combos <- lapply(sn.combos, function(seurat) {AssignAndFilterClusters(seurat)})
```

```{r final clusters, echo=FALSE, fig.height=20, fig.width=12, message=FALSE, warning=FALSE, cache=T}
# standard visuals for clusters and mitocondrial contaminants

cell.plots <- vector("list", length(sn.combos))

for(i in seq(1:length(sn.combos))){
    cell.plots[[i]] <- do_DimPlot(sn.combos[[i]], 
                             reduction = 'umap',
                             label = T,
                             repel = T,
                             shuffle = T) + 
                          NoLegend() + 
                          ggtitle(names(sn.combos)[[i]])

} 

patchwork::wrap_plots(cell.plots,
                          ncol = 2,
                          nrow = 4)


```

```{r music, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# make ExpressionSet w/o outlier 
bulk.es <- ExpressionSet(assayData = as.matrix(bulk[,colnames(bulk) != "wt_Lx3"]))
bulk.es <- exprs(bulk.es)

props <- lapply(sn.combos, function(x){EstimateCellTypeProportions(x, bulk.es, for.aitchison = T)})

```


```{r process props , fig.width = 7, fig.height= 4, echo=FALSE, message=FALSE, warning=FALSE}
# turn music output into graph-friendly dataframe
library(data.table)
decon.melt <- rbindlist(props, fill=TRUE)

decon.melt$CellType = factor(decon.melt$CellType, levels = unique(decon.melt$CellType))

decon.melt$CellType = factor(decon.melt$CellType, levels = unique(decon.melt$CellType))
decon.melt$Prop <- as.numeric(decon.melt$Prop)

# Add genotype info
#! the genotype info assigned here reflects the accurate types
#! genotypes were originally mislabeled when given to us
decon.melt$genotype <- "B6-aKO"
decon.melt[grepl("KO", decon.melt$Sub), 5] <- "B6-WT"
decon.melt[grepl("ko", decon.melt$Sub), 5] <- "B6-WT"
decon.melt$treatment <- "Sham"
decon.melt[grepl("Lx", decon.melt$Sub), 6] <- "Myocardial Infarction"
decon.melt$type <- "whole"
decon.melt[grepl("RNA", decon.melt$Sub), 7] <- "fraction"
decon.melt[grepl("Sham", decon.melt$Sub), 7] <- "fraction"

# Add origin info
decon.melt$origin <- "Jensen"
decon.melt[grepl("B6", decon.melt$Sub), 8] <- "Rau"
decon.melt[grepl("Sham", decon.melt$Sub), 8] <- "Froese"

# Add fraction/CT info
decon.melt$fraction_type <- NA
decon.melt[grepl("CM", decon.melt$Sub), 9] <- "Cardiomyocytes"
decon.melt[grepl("Fib", decon.melt$Sub), 9] <- "Fibroblasts"
decon.melt[grepl("FB", decon.melt$Sub), 9] <- "Fibroblasts"
decon.melt[grepl("Endo", decon.melt$Sub), 9] <- "Endothelial Cells"
decon.melt[grepl("EC_", decon.melt$Sub), 9] <- "Endothelial Cells"

# Add treatment by genotype to make facet wrap clear
decon.melt$gene_treat <- NA
decon.melt$gene_treat <- paste0(
                          decon.melt$genotype, " x ",
                          decon.melt$treatment)     

```

```{r compositional distance, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
library("coda.base")
# make dataframe with expected compositions for each fraction (70 - 95% purity)
# rows = samples , cols = proportions
sim.samples <- bulk.pheno[which(bulk.pheno$fraction != "whole_tissue") ,c(1,3)]
cell.dict <- data.frame(bulk.pheno = unique(sim.samples$fraction),
                        cell.types = c("CMs", "fibroblasts", "endothelial cells"))

sim.fractions <- lapply(sn.combos, function(sn){CreateSimFractions(sn, sim.samples, cell.dict)})

# Function to compare estimated and simulated ("real") compositions of cell type fractions
CalculateAitchisonDistance <- function(sim_fractions, est_fractions) {
  ests <- est_fractions$Est.prop.weighted
  ests <- ests[sim.samples$id,]
  ests[ests == 0] <- 0.05 * 0.65
  aitch_vals <- data.frame(aitchison = rep(NA, dim(ests)[1]))
  rownames(aitch_vals) <- rownames(ests)
  cell_types <- colnames(ests)
  
  for(i in rownames(sim_fractions)){
    aitch_vals[i,] <- coda.base::dist(
      rbind(sim_fractions[i,cell_types],ests[i,cell_types]), 
      method = 'aitchison')[1]
  }

  return(aitch_vals)
}


aitch_vals <- mapply(CalculateAitchisonDistance, sim.fractions, props, SIMPLIFY = FALSE)

cell.names <- bulk.pheno[which(bulk.pheno$fraction != "whole_tissue"),]
```

```{r compositional distance, fig.width = 12, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# Gather aitch_vals list into a dataframe
df <- do.call(rbind, lapply(names(aitch_vals), function(x){
  cbind(data.frame(type=x, aitch_vals[[x]], check.names=FALSE), id = rownames(aitch_vals[[x]]))
}))

# Merge with cell.names
df <- merge(df, cell.names, by = "id", all.x = TRUE)

# Convert to a factor for ordered plotting
df$type <- factor(df$type.x, levels = names(aitch_vals))

# Plot
ggplot(df, aes(x = type, y = aitchison)) +
  geom_boxplot(aes(fill = fraction), width = 0.5, color = "black", outlier.shape = NA) +
  geom_jitter(aes(color = origin, shape = fraction), width = 0.15, size = 3, alpha = 0.7) +
  theme(axis.text.x = element_text(color = "black", size = 15, angle = 30, vjust = 0.6),
        axis.text.y = element_text(color = "black", size = 10),
        axis.ticks.x = element_line(color = "black", size = 1),
        panel.background = element_blank(),
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18),
        axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
        axis.line.y = element_line(color="black", size = 0.5)) +
  labs(x = "Cell Type", y = "Aitchison Value", fill = "Cell Fraction", color = "Origin", shape = "Cell Fraction") +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Set1") +
  scale_shape_manual(values = c(16, 17, 18)) # use different shapes for each fraction

#! you left off here. You still wanted to fix the dots
# they're not aligned with fraction and the colors/shapes are confusing
# also double check that the aitch dists aren't considering factions and cell types twice
# CMs and cardiomyocytes


```
