---
title: "Bisque Comparison"
author: "Brian Gural"
date: "7/5/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Introduction and Context\
The aim of this markdown is to take a look at how different deconvolution packages (MuSiC and Bisque) report compostition. Our single nucleus data is a merged set from two male B6 from Christoph and publically available Tabula Muris data (heart and aorta). Our bulk data is supplied by the Jensen lab, from B6 and a KO line, with heart failure (MI) and untreated (sham). Also, we're including experimentally purified and sequenced homogenous cell fractions (cardiomyocytes, endothelial cells, and fibroblasts). These fractions in duplicate from Christoph and Froese 2022 (doi: 10.1016/j.isci.2022.103965)


```{r load libs, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# load libraries
libs <- c("Seurat", "ggplot2", "DESeq2", "patchwork","SeuratDisk", "MuSiC", "reshape2",
          "tidyverse", "SingleCellExperiment","harmony", "SCpubr","shiny", 
          "AUCell", "viridis", "gplots", "scales", "ggrepel", "gridExtra", "scCustomize",
          "httr","readxl","matrixStats", "TabulaMurisSenisData") # list libraries here
lapply(libs, require, character.only = T)
source("jensen/scripts/functions/decon_all.R")
```


```{r load data, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# bulk
bulk <- read.csv("jensen/data/processed/jensen_rau_froese_cpm", row.names = 1)
bulk_pheno <- read.csv("jensen/data/processed/jensen_rau_froese_pheno", row.names = 1)

# File writtent in jensen/scripts/rmd/doublet_comparison.Rmd

sn <- LoadH5Seurat("jensen/data/processed/single_cell/06282023_NoDoublets.h5seurat")
```

```{r cluster, echo=FALSE,warning=FALSE, cache=TRUE, include=TRUE}
# seurat after 1_1_jensen_doublets.R and 1_2_jensen_merge.R
# should have no doublets and less ambient RNA
# also datasets were filtered before combining for mt percent, features, and counts
sn.clust <- sn |>
    subset(isDoublet == "no") |>
    ClusterSeurat(subset = F, res = 0.05, regress.by = c("origin"))

```

```{r cluster dimplots, echo=FALSE,warning=FALSE, include=TRUE, fig.width= 12, fig.height=18}
plots <- c()
# seurat after 1_1_jensen_doublets.R and 1_2_jensen_merge.R
plots[["dim.clust"]] <- do_DimPlot(sn.clust, label = T, repel = T) + NoLegend()
plots[["dim.origin"]] <- do_DimPlot(sn.clust, group.by = "origin", label = T, repel = T) + NoLegend()

# Convert your table to a data frame and reset row names
df <- as.data.frame(table(sn.clust$origin, sn.clust$seurat_clusters))
colnames(df) <- c("origin", "cluster", "count")

# Convert data frame from wide format to long format

# Convert cluster to factor for ordered x-axis
df$cluster <- factor(df$cluster, levels = unique(df$cluster))

# Create the bar plot
plots[["bar.clust.orig"]] <- ggplot(df, aes(x = cluster, y = count, fill = origin)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(x = "Cluster", y = "Count", fill = "Origin", 
       title = "Count of Origins by Cluster") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

plots[["bar.umi"]] <- VlnPlot(sn.clust, group.by = "origin", features = "nCount_RNA", log = T) 

plots[["feat.count"]] <- do_FeaturePlot(sn.clust, features = "nCount_RNA") 
plots[["feat.features"]] <- do_FeaturePlot(sn.clust, features = "nFeature_RNA") 
plots[["feat.mito"]] <- do_FeaturePlot(sn.clust, features = "PercentMito") 
plots[["feat.doub"]] <- do_FeaturePlot(sn.clust, features = "DoubletScore") 


wrap_plots(plots, ncol = 2)
```



```{r AUCell markers, echo=FALSE,warning=FALSE, include=TRUE, fig.height = 8, fig.width = 12}
# Annotate to cell types
markers.broad <- read.csv("jensen/data/processed/external/mclellan_2020/mclellan_cell_markers_broad.csv")
head(markers.broad)
```

As we talked about last time, AUCell assigns an AUC score (think enrichment) for each cell based on expression and cell types with clear bimodal distributions are retained. The second part, retaining only cell-types with distinct groups, is a change from the last workflow that has been very helpful. Here’s how those AUC distributions look for the cell types we retained:

```{r AUCell plot, echo=FALSE,warning=FALSE, include=TRUE, fig.height = 18, fig.width = 12}
# Annotate to cell types
n_markers <- 25
n_cells <- 15000
 
  seurat <- sn.clust
  cells <- sample(colnames(sn.clust), n_cells, replace = FALSE)
  seurat <- seurat |>
    subset(cells = cells)
  markers.sub <- markers.broad %>%
    group_by(cluster) %>%
    top_n(n_markers, gene) |>
    as.data.frame()
  # Split 'markers' into separate data frames for each unique subcluster
  subclusters <- split(markers.sub, markers.sub$cluster)
  # Extract the gene names from each subcluster data frame
  subcluster_genes <- lapply(subclusters, function(x) x$gene)
  # Create a list with named elements corresponding to each subcluster
  geneSets <- setNames(subcluster_genes, names(subclusters))
  # make expression matrix of single cell
  my.expr <-  seurat@assays$RNA@counts
  # Find  AUC for each cell by each type
  cells_AUC <- AUCell_run(my.expr, geneSets)
  
  cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=F, assign=TRUE)
  
  comments <- lapply(names(cells_assignment), function(cell_type){cells_assignment[[cell_type]]$aucThr$comment})
  good_cells <- names(cells_assignment) #[comments == ""]
  # get the thresholds and report when/how they're passdd
  selectedThresholds <- getThresholdSelected(cells_assignment)
  
  auc.val <- cells_AUC@assays@data@listData$AUC |>
    t()|>
    as.data.frame() 
  auc.val <- auc.val[,good_cells]
  auc.val$seurat <- seurat$seurat_clusters[rownames(auc.val)]

  
# plot 
  

plotAUC <- function(cell_type, auc.df, thresholds){
threshold <- thresholds[[cell_type]]
# Create the histogam
ggplot(auc.df, aes(x = !!sym(cell_type))) +
  geom_histogram(aes(fill = !!sym(cell_type) > threshold), binwidth = 0.01, stat = "bin") +
  scale_fill_manual(values = c("TRUE" = "#D95F02", "FALSE" = "#1B9E77"), guide = FALSE) +
  geom_vline(aes(xintercept = threshold), linetype="dashed", color = "black") +
  theme_minimal() +
  theme(title = element_text(size = 25),
        axis.text = element_text(size = 20)) +
  labs(x = "Value", y = "Count", title = cell_type, subtitle = paste0("Threshold: ", round(threshold, 3))) 
}


plots.auc <- lapply(good_cells, function(x){plotAUC(cell_type = x, auc.df = auc.val, thresholds = selectedThresholds)})
 
wrap_plots(plots.auc, ncol = 2)
```

```{r AUCell heatmap, echo=FALSE,warning=FALSE, include=TRUE, fig.height = 8, fig.width = 12}
# Make correlation matrix for Seurat clusters and cell correlations to cell types

corrs <- model.matrix(~0 + ., data = auc.val) |>
  cor(use = "pairwise.complete.obs") |>
  as.data.frame()
# keep seurat cluster names just as a number
corrs <- corrs[colnames(corrs)[!grepl("seurat", colnames(corrs))], colnames(corrs)[grepl("seurat", colnames(corrs))]] 
colnames(corrs) <- lapply(strsplit(colnames(corrs), "seurat"), "[[", 2)

# format for plotting
corrs <- data.matrix(corrs, rownames.force = T)
empty_columns <- colSums(is.na(corrs) | corrs == "") == nrow(corrs)
standardized_residuals <- corrs[, !empty_columns]

heatmap.2(standardized_residuals, 
          trace = "none", 
          col = viridis(8),
          main = "Correlation coeffiecients between cell-types and clusters",
          xlab = "Seurat Clusters",
          ylab = "Cell Types",
          margins = c(8, 12),
          key.title = "Correlations",
          srtCol = 0,
          cexRow = 1.2)

Idents(sn.clust) <- sn.clust$seurat_clusters

sn.clust <- AssignAndFilterClusters(sn.clust, res.thresh = 0.3, ratio.thresh = 1, min.cell = 200)

```


We set a threshold (0.4 in this case) and assign cell types based on that. Here’s how our final labels look, along with some canonical markers to confirm our annotations:

```{r Cluster Assignments, echo=FALSE,warning=FALSE, include=TRUE, fig.height = 8, fig.width = 16}
plotGene <- function(seurat, slot = "scale.data", gene, plot.title = T, plot.legend = T){
 
 seurat |>
  do_FeaturePlot(gene, reduction = "umap", slot = slot, 
                 legend.position = if(plot.legend == T){"bottom"}else{"none"}, 
                 legend.width = 2, legend.length = 25,  order = T) +
  theme(title = element_text(size = 22),
        legend.title = element_blank()) +
  labs(title = if(plot.title == T){paste0(gene, ", " , names(gene))}else{" "})
}

genes <- c("Pecam1", "Col1a1", "Myh6", "Cd19")
names(genes) <- c("Endothelial Cells", "Fibroblasts", "Cardiomyocytes", "Macrophages/B Cells")

# plot new
plots <- c()
plots[["cell"]] <- do_DimPlot(sn.clust, reduction = "umap", label = T, repel = T, plot.title = "Flexible QC, no doublets")  +  
                                    NoLegend() + theme(title = element_text(size = 22))
plots.exp <- lapply(names(genes), function(x){plotGene(seurat = sn.clust, slot = "data", gene = genes[x], plot.legend = F)})
            
plots.new <- c(plots, plots.exp)


wrap_plots(plots.new, 
           ncol = 3)
```

## Bisque\
```{r DeconvoBuddies Find markers, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# find markers
library(DeconvoBuddies)
library(BisqueRNA)

n_cells <- 10000
cells <- sample(colnames(sn.clust), n_cells, replace = FALSE)

seurat <- sn.clust 
seurat$annots <- Idents(seurat)
seurat$annots[is.na(seurat$annots)] <- "unlabeled"

seurat <- sn.clust |>
    subset(cells = cells, features = rownames(seurat)[rownames(seurat) %in% rownames(bulk)])


  
sn.sce <- as.SingleCellExperiment(seurat, assay = "RNA") |>
  as("SummarizedExperiment")

# add gene names, not sure why they broke

gene_names <- rownames(seurat@assays$RNA@counts)
rownames(sn.sce) <- gene_names

# 
ratios <- get_mean_ratio2(sn.sce, cellType_col = "ident")
fc <- findMarkers_1vAll(sn.sce, cellType_col = "ident", mod = "~orig.ident")
marker_stats <- left_join(ratios, fc, by = c("gene", "cellType.target"))

# narrow markers
marker_stats2 <- marker_stats |>
    filter(gene %in% rownames(sn)) |>
    group_by(cellType.target) |>
    mutate(rank_ratio = row_number())

marker_genes <- marker_stats2 %>%
    filter(rank_ratio <= 25) %>%
    pull(gene)



```


```{r bisque huuki, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}

# Switch to actual data
seurat <- sn.clust |>
    subset(idents = unique(Idents(seurat))[!is.na(unique(Idents(seurat)))])

sn.sce <- as.SingleCellExperiment(seurat, assay = "RNA") 

#### create expression set ####

exp_set_sce <- ExpressionSet(
    assayData = as.matrix(assays(sn.sce)$counts),
    phenoData = AnnotatedDataFrame(
        as.data.frame(colData(sn.sce))[c("ident", "orig.ident")]
    )
)

## Check for nuclei w/ zero expression in marker genes
exp_set_sce <- exp_set_sce[marker_genes, ]
zero_cell_filter <- colSums(exprs(exp_set_sce)) != 0
message("Exclude ", sum(!zero_cell_filter), " cells")
# Exclude 2 cells

exp_set_sce <- exp_set_sce[, zero_cell_filter]

# Christoph fractions + Jensen whole df + Froese fractions, in CPM
jensen_whole <- read.csv("jensen/data/raw/jensen_counts.csv", header = 1, row.names = 1)
rau_fractions <- read.csv("jensen/data/processed/celltype_counts.csv", row.names = 1)

# Merge non-cpm data
all_bulk <- merge(rau_fractions, jensen_whole, by = "row.names")
rownames(all_bulk) <- all_bulk$Row.names
all_bulk <- all_bulk[,-1]

bulk.es <- ExpressionSet(assayData = as.matrix(all_bulk[marker_genes ,colnames(all_bulk) != "wt_Lx3"]))

#### Run Bisque ####
est_prop <- ReferenceBasedDecomposition(
    bulk.eset = bulk.es,
    sc.eset = exp_set_sce,
    cell.types = "ident",
    subject.names = "orig.ident",
    use.overlap = FALSE
)

est_prop$bulk.props <- t(est_prop$bulk.props)


est_prop$Est.prop.long <- est_prop$bulk.props %>%
    as.data.frame() %>%
    tibble::rownames_to_column("Sample") %>%
    pivot_longer(!Sample, names_to = "cell_type", values_to = "prop")


summary(est_prop$bulk.props)
```

```{r deconvolute music , echo=FALSE, warning=FALSE, include=T}

# Load and format bulk data

bulk_pheno <- read.csv("jensen/data/processed/jensen_rau_froese_pheno", row.names = 1)

#bulk.es <- ExpressionSet(assayData = as.matrix(bulk))
bulk.es.exp <- ExpressionSet(assayData = as.matrix(all_bulk)) |>
               exprs()


# Run with all genes and with just subset genes
props.music.markers <- EstimateCellTypeProportions(sn.clust, bulk.es.exp, for.aitchison = F, bulk.log2 = F,
                                                   sn.individuals = "orig.ident", marker = marker_genes, cells_exclude = "endocardium") 
props.music.all <- EstimateCellTypeProportions(sn.clust, bulk.es.exp, for.aitchison = F, 
                                               sn.individuals = "orig.ident", bulk.log2 = F, marker = NULL, cells_exclude = "endocardium")
```

```{r prep phenos , fig.width = 14, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# make one melted DF
props.bisque <- melt(est_prop$bulk.props)
colnames(props.bisque) = c('Sub', 'CellType', 'Prop')
props.bisque$version <- "bisque"
props.music.all$version <- "music.all"
props.music.markers$version <- "music.markers"

decon.melt <- rbind(props.bisque, props.music.all) |>
              rbind(props.music.markers)
# turn music output into graph-friendly dataframe
decon.melt$CellType = factor(decon.melt$CellType, levels = unique(decon.melt$CellType))
decon.melt$Sub = factor(decon.melt$Sub, levels = unique(decon.melt$Sub))
decon.melt$Prop <- as.numeric(decon.melt$Prop)
decon.melt$version = factor(decon.melt$version, levels = unique(decon.melt$version))


# Add genotype info
#! the genotype info assigned here reflects the accurate types
#! genotypes were originally mislabeled when given to us
decon.melt$genotype <- "B6-WT"
decon.melt[grepl("KO", decon.melt$Sub), length(decon.melt)] <- "B6-aKO"
decon.melt[grepl("ko", decon.melt$Sub), length(decon.melt)] <- "B6-aKO"
decon.melt$treatment <- "Sham"
decon.melt[grepl("Lx", decon.melt$Sub), length(decon.melt)] <- "Myocardial Infarction"
decon.melt$type <- "whole"
decon.melt[grepl("RNA", decon.melt$Sub), length(decon.melt)] <- "fraction"
decon.melt[grepl("Sham", decon.melt$Sub), length(decon.melt)] <- "fraction"

# Add origin info
decon.melt$origin <- "Jensen"
decon.melt[grepl("B6", decon.melt$Sub), length(decon.melt)] <- "Rau"
decon.melt[grepl("Sham", decon.melt$Sub), length(decon.melt)] <- "Froese"

# Add fraction/CT info
decon.melt$fraction_type <- NA
decon.melt[grepl("CM", decon.melt$Sub), length(decon.melt)] <- "Cardiomyocytes"
decon.melt[grepl("Fib", decon.melt$Sub), length(decon.melt)] <- "Fibroblasts"
decon.melt[grepl("FB", decon.melt$Sub), length(decon.melt)] <- "Fibroblasts"
decon.melt[grepl("Endo", decon.melt$Sub), length(decon.melt)] <- "Endothelial Cells"
decon.melt[grepl("EC_", decon.melt$Sub), length(decon.melt)] <- "Endothelial Cells"

# Add treatment by genotype to make facet wrap clear
decon.melt$gene_treat <- NA
decon.melt$gene_treat <- paste0(
                          decon.melt$genotype, " x ",
                          decon.melt$treatment)  



```




```{r music plot controls, fig.width = 12, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
plotDeconFractions <- function(x){decon.melt |>
  subset(type == "fraction" & version == x) |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 0.8,
           color = "black")+
   scale_fill_brewer(name = "Cell Type",
                    palette = "Dark2") +
   facet_wrap(~fraction_type, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
   ylab("Proportion") +
   theme_minimal() +
   theme( axis.text.x = element_blank(),
         strip.text = element_text(size = 20),
         title = element_text(size = 20),
         legend.text = element_text(size = 18),
         axis.ticks.x = element_blank()) +
   xlab("Samples") +
    ggtitle(x)
}

versions <- decon.melt$version |>
              unique()

versions <- c("music.all", "music.markers")
fraction.plots <- lapply(versions, function(x){plotDeconFractions(x)})

wrap_plots(fraction.plots, ncol = 1)


```

## Composition estimates in Jensen data

Here’s the actual results you’re intereted in. The KO vs. WT labels were based on the ‘id’ column of the file you gave me (labels include WT5, AKO5, Ako_Lx1, wt_Lx1, and Ako_Lx4), rather than the ‘condition’ column (labels include: WT.sham, KO.sham, WT.MI, KO.MI, and WT.outlier). Just let me know if I got this right, since I know you had mentioned that they were switched.

```{r music plot whole, fig.width = 12, fig.height= 16, echo=FALSE, message=FALSE, warning=FALSE}
plotDeconWhole <- function(x){decon.melt |>
  subset(type == "whole" & version == x) |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 0.8,
           color = "black")+
   scale_fill_brewer(name = "Cell Type",
                    palette = "Dark2") +
   facet_wrap(~gene_treat, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
   ylab("Proportion") +
   theme_minimal() +
   theme( axis.text.x = element_blank(),
         strip.text = element_text(size = 20),
         title = element_text(size = 20),
         legend.text = element_text(size = 18),
         axis.ticks.x = element_blank()) +
   xlab("Samples") +
    ggtitle(x)
}

versions <- decon.melt$version |>
              unique()

versions <- c("music.all", "music.markers")
fraction.plots <- lapply(versions, function(x){plotDeconWhole(x)})

wrap_plots(fraction.plots, ncol = 1)

```

```{r music plot sham, fig.width = 12, fig.height= 12, echo=FALSE, message=FALSE, warning=FALSE}
decon.melt$CellType <- reorder(decon.melt$CellType, -decon.melt$Prop, FUN = sum)

plots <- c()
plots[[1]]  <- decon.melt |>
  subset(type == "whole" & treatment == "Sham" & version == "music.markers") |>
ggplot(aes(x = CellType, y = Prop, fill = genotype)) +
  geom_bar(stat = "summary", 
           position = position_dodge(0.9), 
           fun = mean,
           width = 0.9,
           color = "black") +
  theme(axis.text.x = element_text(color = "black", size = 15, angle =15), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks = element_blank(), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  geom_jitter(inherit.aes = T, 
              position = position_jitterdodge(0.01, 0, 0.9),
              size = 2) +
  facet_zoom(ylim = c(0, 0.2), zoom.data = ifelse(a <= 0.3, NA, FALSE)) +
  ggtitle("WT vs KO at baseline (sham surgery)")


p


plots[[2]] <- decon.melt |>
  subset(type == "whole" & treatment == "Myocardial Infarction" & version == "music.markers") |>
ggplot(aes(x = CellType, y = Prop, fill = genotype)) +
  geom_bar(stat = "summary", 
           position = position_dodge(0.9), 
           fun = mean,
           width = 0.9,
           color = "black") +
  theme(axis.text.x = element_text(color = "black", size = 15, angle =15), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks = element_blank(), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  geom_jitter(inherit.aes = T, 
              position = position_jitterdodge(0.01, 0, 0.9),
              size = 2) +
  facet_zoom(ylim = c(0, 0.2), zoom.data = ifelse(a <= 0.3, NA, FALSE)) +
  ggtitle("WT vs KO during myocardial infarction")

q
wrap_plots(plots, ncol = 1)
```

```{r dirichlet, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
library("DirichletReg")
# prep DirichletReg matrix
dir.cols <- colnames(decon.melt)[c(1:3,5,6)]

dir.mat <- decon.melt |>
  subset(type == "whole" & version == "music.markers",
         select = dir.cols) |>
  dcast(Sub + genotype + treatment ~ CellType, value.var = "Prop")

# Convert data to DirichletRegData object
dir.mat$CellTypes <- DR_data(dir.mat[,c(4:length(dir.mat))])

# Run Dirichlet regression
model.1 <- DirichReg(CellTypes ~ treatment + genotype + treatment * genotype, data = dir.mat)

model.2 <- DirichReg(CellTypes ~ treatment + genotype, data = dir.mat)

# compare models, find if interaction term improves model
anova(model.1, model.2)

# Pr(>Chi) is highly significant (p = 2.804e-07), which means there's strong evidence against the null hypothesis (the simple model is better). model.1 provides a significantly better fit to the data than model.2.
# this implies that the effect of the treatment may be different for different genotypes.
summary(model.1)

write.csv(decon.melt, "jensen/results/estimates/deconbuddies_marker_test_07072023")
```