---
title: "Integrate sn and sc by downsampling"
author: "Brian Gural"
date: "6/13/2023"
output:
  html_document:
    toc: true
    toc_float: true
---




```{r setup, include=FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Introduction and Context\
The aim of this analysis is to estimate the cellular composition of bulk-RNAseq reads from mouse left ventricular tissue. We'll find cell-type-specific gene expression reference profiles with snRNAseq. These two single nucleus samples are aggregates of several adult, male B6 from Christoph's post-doc.\
\
This document explains the workflow in the order in which the data processing occurred. You can jump to portions with the table of contents on the left.\

## Load data and packages\


```{r load libs, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# load libraries
libs <- c("Seurat", "ggplot2", "DESeq2", "patchwork","SeuratDisk", "MuSiC", "reshape2",
          "tidyverse", "SingleCellExperiment","harmony", "SCpubr","shiny", 
          "AUCell", "viridis", "gplots", "scales", "ggrepel", "gridExtra", "scCustomize",
          "httr","readxl","matrixStats", "TabulaMurisSenisData") # list libraries here
lapply(libs, require, character.only = T)
source("jensen/scripts/functions/decon_all.R")
```

```{r load data, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# bulk
bulk <- read.csv("jensen/data/processed/jensen_rau_froese_cpm", row.names = 1)
bulk_pheno <- read.csv("jensen/data/processed/jensen_rau_froese_pheno", row.names = 1)

# Single cell
sn.rau <- LoadH5Seurat("jensen/data/raw/b6_sn.h5seurat")
sn.tm  <- LoadH5Seurat("jensen/data/processed/tabula_muris.h5seurat")
sn.wu  <- LoadH5Seurat("jensen/data/processed/single_cell/wu_2021.h5seurat")
sn.mt  <- LoadH5Seurat("jensen/data/processed/single_cell/martini_2019.h5seurat")

sn.rau$origin <- "rau"

sn.all <- merge(sn.rau, sn.tm) |>
            merge(sn.wu) |>
            merge(sn.mt)

sn.all$PercentMito <- PercentageFeatureSet(sn.all, pattern = "^mt-")
```

```{r downsample, message=FALSE, warning=FALSE, cache=FALSE, include=F}

sn.small <- sn.all |>
  GetAssayData(assay = "RNA", slot = "counts") |>
  as.matrix() |>
  SampleUMI(5000, verbose = T)

sn.small <- CreateSeuratObject(sn.small)

meta.features <- colnames(sn.all@meta.data) [-c(2,3)]
for(i in meta.features){
  sn.small <- AddMetaData(sn.small, sn.all@meta.data[[i]], col.name = i)
}

```


```{r cluster 1, message=FALSE, warning=FALSE, cache=FALSE, include=F}
# sn.clust.1 <- ClusterSeurat(sn.small, max.rna.ft = 5000,
#                            max.mt.pt = 5, res = 0.03, regress.by = c("origin"))
sn.clust.1 <- ClusterSeurat(sn.all, max.rna.ft = 25000,
                            max.mt.pt = 5, res = 0.03, regress.by = c("origin"))
```

```{r cluster1.dim, echo=FALSE, fig.height=14, fig.width=10, message=FALSE, warning=FALSE, cache=T}
# standard visuals for clusters and mitocondrial contaminants
dim.ft  <- c("seurat_clusters", "origin", "orig.ident")
feat.ft <- c("PercentMito", "nCount_RNA")
vln.ft <- c("PercentMito")
vln.groups <- c("seurat_clusters")

plotUMAP(data = sn.clust.1,
         dim.ft = dim.ft,
         feat.ft = feat.ft,
         nrow = 3, 
         ncol = 2,
         font.size = 12)



```



```{r load broad markers, echo=F, fig.height=10, fig.width=8, message=FALSE, warning=FALSE, cache=TRUE}
# load markers
sn.clust.2 <- sn.clust.1
markers.broad <- read.csv("jensen/data/processed/external/mclellan_2020/mclellan_cell_markers_broad.csv")
```

\
```{r broad.cell.types AUC, echo=FALSE, fig.height=4, fig.width=7, message=FALSE, warning=FALSE, cache=TRUE}
# Split 'markers' into separate data frames for each unique subcluster
subclusters <- split(markers.broad, markers.broad$cluster)
# Extract the gene names from each subcluster data frame
subcluster_genes <- lapply(subclusters, function(x) x$gene)
# Create a list with named elements corresponding to each subcluster
geneSets <- setNames(subcluster_genes, names(subclusters))
# make expression matrix of single cell
my.expr <-  sn.clust.2@assays$RNA@counts
# Find  AUC for each cell by each type
cells_AUC <- AUCell_run(my.expr, geneSets)
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=T, assign=TRUE)


# select thresholds only if they have a biomodal distribution threshold
cells <- subclusters |>
          names() |>
          unique()

# Initialize df
thresholds <- data.frame(threshold <- rep(NA, length(cells)))
row.names(thresholds) <- cells

for(i in cells){
  cell.info <- cells_assignment[[i]]
  thresh.type <- names(cell.info$aucThr$selected)
  thresh.val <- cell.info$aucThr$selected
  
  if(thresh.type == "minimumDens"){
    thresholds[i,1] <- thresh.val[[1]]
  }
}
# get the thresholds and report when/how they're passed
selectedThresholds <- getThresholdSelected(cells_assignment)

selectedThresholds["cardiomyocytes"] <- 0.25

auc.val <- cells_AUC@assays@data@listData$AUC |>
              as.data.frame() |>
              t()

# Apply process_cell_id function to each row of auc.val using lapply()
broad_cell_types_list <- lapply(1:nrow(auc.val), function(i) processCellId(auc.val[i, ]))

# Convert the list to a data frame
broad.cell.types <- data.frame(BroadCellType = unlist(broad_cell_types_list), stringsAsFactors = FALSE)
rownames(broad.cell.types) <- rownames(auc.val)

sn.clust.2 <- AddMetaData(sn.clust.2, broad.cell.types)
```


### Label Clusters\

```{r dim broad/fine, echo=FALSE, fig.height=18, fig.width=10, message=FALSE, warning=FALSE,}
# standard visuals for clusters and mitocondrial contaminants
dim.ft  <- c("BroadCellType", "seurat_clusters")
feat.ft <- "PercentMito"
plotUMAP(data = sn.clust.2,
         dim.ft = dim.ft,
         feat.ft = feat.ft,
         nrow = 3, 
         ncol = 1,
         font.size = 8)

```
\

### Defining Seurat clusters\


```{r heatmap for all celltypes, echo=FALSE, fig.height=12, fig.width=10, message=FALSE, warning=FALSE, eval=T}
# Create a contingency table
contingency_table <- table(c(sn.clust.2$BroadCellType),
                           c(sn.clust.2$seurat_clusters), 
                           useNA = "ifany")

# Perform the chi-squared test
chi_squared_test <- chisq.test(contingency_table)

# Calculate the standardized residuals
observed_frequencies <- contingency_table
expected_frequencies <- chi_squared_test$expected
standardized_residuals <- (observed_frequencies - expected_frequencies) / sqrt(expected_frequencies)

rownames(standardized_residuals)[is.na(rownames(standardized_residuals))] <- "unlabeled"

col_palette <- viridis(25)
# Create a heatmap of the standardized residuals
heatmap.2(standardized_residuals, 
          trace = "none", 
          col = col_palette,
          main = "Residuals, All Cell Types",
          xlab = "Seurat Clusters",
          ylab = "All Cell Types",
          margins = c(8, 12),
          key.title = "Standardized Residuals",
          srtCol = 0,
          cexRow = 1.2)

```
\

```{r label clusters w chisq for celltypes, echo=FALSE, fig.height=6, fig.width=10, message=FALSE, warning=FALSE}

# exclude clusters below specific max residual threshold
res.thresh <- 80
max <- apply(standardized_residuals, 2, max)
max.exclude <- max < res.thresh

# exclude clusters whose max residual is less than double the unlabelled residual
ratio.thresh <- 2
unlabel.exclude <- abs(max/standardized_residuals[nrow(standardized_residuals),]) < ratio.thresh

# exclude clusters with few cells 

# remove very small clusters 
bad.clusts <- table(Idents(sn.clust.2))[table(Idents(sn.clust.2)) < 400]

# get final exclusion list
all.exclude <- unlabel.exclude | unlabel.exclude
# Find the index of the largest value in each column of the normalized_contingency_table
max.indices <- apply(standardized_residuals, 2, which.max)

max.indices[all.exclude] <- NA
# exclude clusters below specific max residual threshold

# Get the corresponding BroadCellType for each Seurat cluster
assigned.cell.types <- rownames(standardized_residuals)[max.indices] 

#rename the clusters
names(assigned.cell.types) <- levels(sn.clust.2)
sn.clust.2 <- RenameIdents(sn.clust.2, assigned.cell.types)

# remove very small clusters 
bad.clusts <- names(table(Idents(sn.clust.2)))[table(Idents(sn.clust.2)) < 400]

Idents(sn.clust.2)[which(Idents(sn.clust.2) %in% bad.clusts)] <- NA
#SaveH5Seurat(sn.clust.2, "jensen/data/processed/single_cell/06062023")

```

Here's what our final clusters and labels look like: \
```{r final cluster labels, echo=FALSE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE}
do_DimPlot(sn.clust.2, reduction = "umap", label = T, repel = T) + NoLegend()
```

```{r  cluster origin, echo=FALSE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE}
do_DimPlot(sn.clust.2, reduction = "umap",group.by = "origin", label = T, repel = T, shuffle = T) + NoLegend()
```

## MuSiC and composition estimates\


```{r cluster2.dim, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# music_prop needs exprs matrix
sn.sce <- as.SingleCellExperiment(sn.clust.2, assay = "RNA")

# make ExpressionSet w/o outlier 
bulk.es <- ExpressionSet(assayData = as.matrix(bulk[,colnames(bulk) != "wt_Lx3"]))
bulk.es <- exprs(bulk.es)


# exclude unlabeled cluster
cells <- levels(Idents(sn.clust.2))
cells[cells != "unlabeled"]
# Access the cell type using the index

decon <- music_prop(bulk.mtx = 2^bulk.es, sc.sce = sn.sce, markers = NULL,
                    clusters = "ident", samples = "orig.ident",
                    select.ct = cells)
```


```{r cluster2.feat , fig.width = 7, fig.height= 4, echo=FALSE, message=FALSE, warning=FALSE}
# turn music output into graph-friendly dataframe
decon.melt = melt(decon$Est.prop.weighted)
colnames(decon.melt) = c('Sub', 'CellType', 'Prop')
decon.melt$CellType = factor(decon.melt$CellType, levels = unique(decon.melt$CellType))

decon.melt$CellType = factor(decon.melt$CellType, levels = unique(decon.melt$CellType))
decon.melt$Prop <- as.numeric(decon.melt$Prop)

# Add genotype info
#! the genotype info assigned here reflects the accurate types
#! genotypes were originally mislabeled when given to us
decon.melt$genotype <- "B6-aKO"
decon.melt[grepl("KO", decon.melt$Sub), 4] <- "B6-WT"
decon.melt[grepl("ko", decon.melt$Sub), 4] <- "B6-WT"
decon.melt$treatment <- "Sham"
decon.melt[grepl("Lx", decon.melt$Sub), 5] <- "Myocardial Infarction"
decon.melt$type <- "whole"
decon.melt[grepl("RNA", decon.melt$Sub), 6] <- "fraction"
decon.melt[grepl("Sham", decon.melt$Sub), 6] <- "fraction"

# Add origin info
decon.melt$origin <- "Jensen"
decon.melt[grepl("B6", decon.melt$Sub), 7] <- "Rau"
decon.melt[grepl("Sham", decon.melt$Sub), 7] <- "Froese"

# Add fraction/CT info
decon.melt$fraction_type <- NA
decon.melt[grepl("CM", decon.melt$Sub), 8] <- "Cardiomyocytes"
decon.melt[grepl("Fib", decon.melt$Sub), 8] <- "Fibroblasts"
decon.melt[grepl("FB", decon.melt$Sub), 8] <- "Fibroblasts"
decon.melt[grepl("Endo", decon.melt$Sub), 8] <- "Endothelial Cells"
decon.melt[grepl("EC_", decon.melt$Sub), 8] <- "Endothelial Cells"

# Add treatment by genotype to make facet wrap clear
decon.melt$gene_treat <- NA
decon.melt$gene_treat <- paste0(
                          decon.melt$genotype, " x ",
                          decon.melt$treatment)     

```

```{r music plot controls, fig.width = 12, fig.height= 7, echo=FALSE, message=FALSE, warning=FALSE}
decon.melt |>
  subset(type == "fraction") |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 0.5,
           color = "black")+
   scale_fill_brewer(name = "Cell Type",
                    palette = "Dark2") +
   facet_wrap(~fraction_type, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
   ylab("Proportion") +
   theme( axis.text.x = element_text(color = "black", size = 8, angle = 30),
         strip.text = element_text(size = 20),
         title = element_text(size = 20),
         legend.text = element_text(size = 18),
         axis.ticks.x = element_blank()) +
   xlab("Samples (All UMI)")

```
```{r music plot whole, fig.width = 12, fig.height= 7, echo=FALSE, message=FALSE, warning=FALSE}
decon.melt |>
  subset(type == "whole") |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 0.5,
           color = "black")+
   scale_fill_brewer(name = "Cell Type",
                    palette = "Dark2") +
   facet_wrap(~gene_treat, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
   ylab("Proportion") +
   theme( axis.text.x = element_text(color = "black", size = 8, angle = 30),
         strip.text = element_text(size = 20),
         title = element_text(size = 20),
         legend.text = element_text(size = 18),
         axis.ticks.x = element_blank()) +
   xlab("Samples (All UMI)")

```
```{r music plot simonson sham, fig.width = 12, fig.height= 7, echo=FALSE, message=FALSE, warning=FALSE}
library(ggbreak)
library(ggplot2)
library(ggforce)
decon.sham <- decon.melt |>
  subset(type == "whole" & treatment == "Sham")
decon.sham$CellType <- reorder(decon.sham$CellType, -decon.sham$Prop, FUN = sum)

#colorbrewer2.org
colors.light <- c("B6-aKO" = "#b8e186", "B6-WT" = "#f1b6da")
colors.dark <- c("#4d9221", "#c51b7d")
# Create a grouped bar plot
z <- decon.sham |>
  subset(type == "whole" & treatment == "Sham") |>
ggplot(aes(x = CellType, y = Prop, fill = genotype)) +
  geom_bar(stat = "summary", 
           position = position_dodge(0.9), 
           fun = mean,
           width = 0.9,
           color = "black") +
  theme(axis.text.x = element_text(color = "black", size = 15, angle = 30), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks = element_blank(), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  geom_jitter(inherit.aes = T, 
              position = position_jitterdodge(0.01, 0, 0.9),
              size = 2) +
  scale_y_break(c(0.08,0.92))

z
p <- decon.sham |>
  subset(type == "whole" & treatment == "Sham") |>
ggplot(aes(x = CellType, y = Prop, fill = genotype)) +
  geom_bar(stat = "summary", 
           position = position_dodge(0.9), 
           fun = mean,
           width = 0.9,
           color = "black") +
  theme(axis.text.x = element_text(color = "black", size = 15, angle = 30), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks = element_blank(), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  geom_jitter(inherit.aes = T, 
              position = position_jitterdodge(0.01, 0, 0.9),
              size = 2) +
  facet_zoom(ylim = c(0, 0.07), zoom.data = ifelse(a <= 0.07, NA, FALSE))

p
```
```{r music plot simonson mi, fig.width = 12, fig.height= 7, echo=FALSE, message=FALSE, warning=FALSE}
library(ggbreak)
library(ggplot2)
library(ggforce)
decon.mi <- decon.melt |>
  subset(type == "whole" & treatment == "Myocardial Infarction")
decon.mi$CellType <- reorder(decon.mi$CellType, -decon.mi$Prop, FUN = sum)

ggplot(decon.mi, 
          aes(x = CellType, y = Prop, fill = genotype)) +
        geom_bar(stat = "summary", 
           position = position_dodge(0.9), 
           fun = mean,
           width = 0.9,
           color = "black") +
        theme(axis.text.x = element_text(color = "black", size = 15, angle = 30, vjust = 0.6), # Increase size here for larger axis text
          axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
          axis.ticks.x = element_line(color = "black", size = 1), # Add this line for thick black axis bars
          legend.position = c(.95, .95),
          legend.justification = c("right", "top"),
          legend.box.just = "right",
          legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
          panel.background = element_blank(), # Remove background gray color
          strip.text = element_text(size = 20),
          title = element_text(size = 20),
          legend.text = element_text(size = 18),
          axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
          axis.line.y = element_line(color="black", size = 0.5)) +
        labs(x = "Cell Type", 
          y = "Proportion", 
          fill = "Genotype") +
        scale_fill_brewer(palette = "Dark2") + 
        geom_jitter(inherit.aes = T, 
          position = position_jitterdodge(0.01, 0, 0.9),
              size = 2) + 
        facet_zoom(ylim = c(0, 0.2))
#+scale_y_break(c(0.2,0.7))
```
## DirichletReg\

```{r DirichletReg, eval=FALSE, fig.height=8, fig.width=7, message=FALSE, warning=FALSE, include=FALSE}

library("DirichletReg")
library("Ternary")
library(reshape2)
# prep DirichletReg matrix
dir.cols <- colnames(decon.melt)[c(1:3,5,6)]

dir.mat <- decon.melt |>
  subset(type == "whole" &
         CellType != "B cells",
         select = dir.cols) |>
  dcast(Sub + genotype + treatment ~ CellType, value.var = "Prop")



# Convert data to DirichletRegData object
dir.mat$CellTypes <- DR_data(dir.mat[,c(4:6)])

# Run Dirichlet regression
model.1 <- DirichReg(CellTypes ~ treatment + genotype + treatment * genotype, data = dir.mat)

model.2 <- DirichReg(CellTypes ~ treatment + genotype, data = dir.mat)

# compare models, find if interaction term improves model
anova(model.1, model.2)

# Pr(>Chi) is highly significant (p = 8.573e-05), which means there's strong evidence against the null hypothesis (the simple model is better). model.1 provides a significantly better fit to the data than model.2.
# this implies that the effect of the treatment may be different for different genotypes.
summary(model.1)

SaveH5Seurat(sn.clust.2, "jensen/data/processed/single_cell/061502023")
```