---
title: "Jolene Chat"
author: "Brian Gural"
date: "6/28/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Project Goals\
Estimate cellular composition in mammalian hearts. I'm using sn/scRNAseq as a reference for deconvolution of bulk RNAseq. For this walkthrough, I'm going to cover a collaboration I'm doing with Brian Jensen. 

### Jensen Collab\
They're interested in alpha-1 adronergic receptor as a mediator of heart failure (HF). He asked me to estimate composition of his bulk RNAseq from wild type and KO mice with and without HF. I'm using a combination of sn/sc references. One of these is from Christoph, but all of the others are publically available.

### Overview of Workflow:

1. Load the data and save it 
    - Currently four sources totalling to ~70k droplets
    - Data downloaded as GSM from GEO
    - QC'd post-10x prior to receiving
2. QC and filtering
    - Remove doublets and ambient RNA with dropletUtils 
      - ? I would need the raw matricies from SRA to do this correctly?
    - Filter each dataset by quantiles of counts, features, and mitocondrial gene percent
3. Merge datasets and cluster
    - Integrate with Harmony
      - Currently correcting by dataset origin
4. Label and exclude clusters
    - Use AUC and known markers to annotate cells
      - Labelling based on bimodal distributions from all datasets integrated
      - Using AUCell and markers from McKellan 2020
    - Downsampling cell to save time (currently 10K out of 33K)
    - Label clusters with cell types
      - Correlate scores of cells to assigned clusters
    - Exclude clusters based on:
      - <200 cells
      - Corr between seurat cluster and label > 0.4
5. Deconvolute with MuSiC
    - Automatic marker selection
    - Deconvolution of fractions and whole tissue samples
6. Compare compositional distances
    - Measure deconvolution performance with Aitchinson distance between:
      - Fraction composition estimates from MuSiC outputs
      - Simulated composition based on likely purity of our ground truths
    - Model compositional shifts with Dirichlet regression
      

### Comparisons I've tried

1. Effect of downsampling UMI per droplet to 5K
    - Asking if variation in average counts by dataset effects results
    - Not much difference
    - Run before doublet/ambient removal
2. Results with permutations of every dataset
    - More datasets generally perform better on fraction deconvolution 
    - Run before doublet/ambient removal
3. Effect of doublet/ambient RNA removal and flexible filtering of datasets
    - Flexible filtering is better
    - Doublet removal doesn't do all that much


### Visuals:

#### Clustering/QC
1. VlnPlots of datasets of origin (counts/features)
2. barplot of total umi per dataset
3. ?Dimplot of sn.clust.new w/ and w/o harmony
4. Dimplots post clustering of clusters, origin, cell types
5. Features (counts, mito, features, doublet score)

#### Ground truth comparisons
1. composition barplot from sn.clust.new
2. Aitchison plots of dataset combos and doublet filtering


#### Composition estimates
1. four panel geno x treat plots
2. zoomed plots of geno comparison in sham and mi
3. Dirilchet model code and outputs

#### Cell type annotations
1. AUC plot example (good and bad)
2. Feature plots of good cell types
3. Heatmap of cluster-label correlations
4. final cell type labels
5. Expression of canonical markers plot

```{r load libs, message=FALSE, warning=FALSE, cache=TRUE, include=F}
# load libraries
libs <- c("Seurat", "ggplot2", "DESeq2", "patchwork","SeuratDisk", "MuSiC", "reshape2",
          "tidyverse", "SingleCellExperiment","harmony", "SCpubr","shiny", 
          "AUCell", "viridis", "gplots", "scales", "ggrepel", "gridExtra", "scCustomize",
          "httr","readxl","matrixStats", "TabulaMurisSenisData") # list libraries here
lapply(libs, require, character.only = T)
source("jensen/scripts/functions/decon_all.R")
```

```{r load data, message=FALSE, warning=FALSE, cache=TRUE, include=TRUE}
# bulk
bulk <- read.csv("jensen/data/processed/jensen_rau_froese_cpm", row.names = 1)
bulk_pheno <- read.csv("jensen/data/processed/jensen_rau_froese_pheno", row.names = 1)

# Single cell of all datasets combined
sn <- LoadH5Seurat("jensen/data/processed/single_cell/06282023_NoDoublets.h5seurat")
print(sn)
```

```{r cluster, echo=FALSE,warning=FALSE, cache=TRUE, include=TRUE}
# seurat after 1_1_jensen_doublets.R and 1_2_jensen_merge.R
# should have no doublets and less ambient RNA
# also datasets were filtered before combining for mt percent, features, and counts,
sn.clust <- sn |>
    subset(isDoublet == "no") |>
    ClusterSeurat(subset = F, res = 0.05, regress.by = c("origin"))

```

```{r cluster dimplots, echo=FALSE,warning=FALSE, include=TRUE, fig.width= 12, fig.height=18}
plots <- c()
# seurat after 1_1_jensen_doublets.R and 1_2_jensen_merge.R
plots[["dim.clust"]] <- do_DimPlot(sn.clust, label = T, repel = T) + NoLegend()
plots[["dim.origin"]] <- do_DimPlot(sn.clust, group.by = "origin", label = T, repel = T) + NoLegend()
plots[["feat.count"]] <- do_FeaturePlot(sn.clust, features = "nCount_RNA") 
plots[["feat.features"]] <- do_FeaturePlot(sn.clust, features = "nFeature_RNA") 
plots[["feat.mito"]] <- do_FeaturePlot(sn.clust, features = "PercentMito") 
plots[["feat.doub"]] <- do_FeaturePlot(sn.clust, features = "DoubletScore") 
plots[["bar.umi"]] <- VlnPlot(sn.clust, group.by = "origin", features = "nCount_RNA", log = T) 

# Convert your table to a data frame and reset row names
df <- as.data.frame(table(sn.clust$origin, sn.clust$seurat_clusters))
colnames(df) <- c("origin", "cluster", "count")

# Convert data frame from wide format to long format

# Convert cluster to factor for ordered x-axis
df$cluster <- factor(df$cluster, levels = unique(df$cluster))

# Create the bar plot
plots[["bar.clust.orig"]] <- ggplot(df, aes(x = cluster, y = count, fill = origin)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(x = "Cluster", y = "Count", fill = "Origin", 
       title = "Count of Origins by Cluster") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

wrap_plots(plots, ncol = 2)
```


```{r AUCell new, echo=FALSE,warning=FALSE, include=TRUE, fig.height = 8, fig.width = 12}
# Annotate to cell types
markers.broad <- read.csv("jensen/data/processed/external/mclellan_2020/mclellan_cell_markers_broad.csv")

n_markers <- 15
# get AUC value of a representative subset of cells for cell marker enrichment
auc.new <- AssignAnnotations(sn.clust.new, markers = markers.broad, n_cores = 1, n_markers = n_markers, n_cells = 20000)

# Make correlation matrix for Seurat clusters and cell correlations to cell types

corrs <- model.matrix(~0 + ., data = auc.new) |>
  cor(use = "pairwise.complete.obs") |>
  as.data.frame()
# keep seurat cluster names just as a number
corrs <- corrs[colnames(corrs)[!grepl("seurat", colnames(corrs))], colnames(corrs)[grepl("seurat", colnames(corrs))]] 
colnames(corrs) <- lapply(strsplit(colnames(corrs), "seurat"), "[[", 2)

# format for plotting
corrs <- data.matrix(corrs, rownames.force = T)
empty_columns <- colSums(is.na(corrs) | corrs == "") == nrow(corrs)
standardized_residuals <- corrs[, !empty_columns]

heatmap.2(standardized_residuals, 
          trace = "none", 
          col = col_palette,
          main = "Residuals without doublets",
          xlab = "Seurat Clusters",
          ylab = "Cell Types",
          margins = c(8, 12),
          key.title = "Correlations",
          srtCol = 0,
          cexRow = 1.2)

sn.clust.new.1 <- AssignAndFilterClusters(sn.clust, res.thresh = 0.4, ratio.thresh = 1.3, min.cell = 200)

#sn.clust.old.1 <- AssignAnnotations(sn.clust.old, markers = markers.broad, n_cores = 1, n_markers = 50)
#sn.clust.uniform.1 <- AssignAnnotations(sn.clust.uniform, markers = markers.broad, n_cores = 4, n_markers = 50)
```
Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
