---
title: "Deconvolution after ambient RNA removal"
author: "Brian Gural"
date: "2023-08-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 10, message = F, warning = F, error = T)
```

```{r load libs, message=FALSE, echo = T, warning=FALSE, results= 'hide', cache=FALSE}
libs <- c("Seurat", "ggplot2", "DESeq2", "patchwork","SeuratDisk", "MuSiC", "reshape2",
          "tidyverse", "SingleCellExperiment", "harmony", "SCpubr", 
          "AUCell", "viridis", "gplots", "scales", "ggrepel", "gridExtra", "scCustomize",
          "httr","matrixStats", "scran", "scuttle", "scater", "DropletUtils", "scDblFinder",
          "DeconvoBuddies", "readr", "SingleCellExperiment", "compositions",
          "SummarizedExperiment", "Biobase", "ggmagnify", "stringr", "MuSiC",
          "ggforce", "DirichletReg", "readxl") # list libraries here
lapply(libs, require, character.only = T)
source("jensen/scripts/functions/decon_all.R")
rm(libs)
```


```{r load data, message=FALSE, echo = T, warning = FALSE, cache = T, collapse = T}

sn <- LoadH5Seurat("jensen/data/processed/single_cell/merged_datasets/08162023.h5seurat")

# Load bulk data
bulk <- read_csv("jensen/data/processed/jensen_rau_froese_cpm") |>
  column_to_rownames(var = "...1")

# Load and format bulk data
jensen_bulk <- readxl::read_xlsx("jensen/data/raw/jensen_counts_correct.xlsx")[,-c(20,21)] |>
  as.data.frame()
colnames(jensen_bulk)[c(20,21)] <- c("WT MI", "KO MI")

```


```{r DeconvoBuddies Find markers, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# Update sn with annotations
sn <- sn |>
         subset(features = rownames(sn)[rownames(sn) %in% rownames(bulk)]) 

sn.sce <- sn|> 
  as.SingleCellExperiment(assay = "RNA") |>
  as("SummarizedExperiment")

# Add gene names
rownames(sn.sce) <- rownames(sn@assays$RNA@counts)

# Get markers based on expression ratios between clusters and 1vAll comparisons

scn.markers  <- scran::findMarkers(sn.sce, groups = Idents(sn), pval.type = "all", direction = "up")


.getMarkers <- function(type){
marker <- scn.markers@listData[[type]] |>
  as.data.frame() |> 
  subset(p.value == 0) |>
  arrange(desc(summary.logFC)) |>
  slice_head(n = 25) |>
  select(p.value, FDR, summary.logFC) |>
  mutate(celltype = type) |> 
  rownames_to_column(var = "gene")
return(marker)}


all.markers <- lapply(levels(Idents(sn)), function(x){.getMarkers(x)}) |> 
               reduce(full_join)

all.markers |>
  group_by(celltype) |>
  arrange(desc(summary.logFC)) |>
  slice_head(n = 3) 

gc()
```

```{r deconvolute music , echo=FALSE, warning=FALSE, include=T}
# Prep bulk matrix as ExpressionSet
bulk.es.exp <- bulk |> 
  as.matrix() %>%
  ExpressionSet(assayData = .) |> 
  exprs()

# Run MuSiC with subset genes
decon.melt <- EstimateCellTypeProportions(sn, bulk.es.exp, for.aitchison = FALSE, bulk.log2 = TRUE,
                                          sn.individuals = "orig.ident", marker = all.markers$gene, cells_exclude = NULL)
 
```

```{r prep phenos , fig.width = 14, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
phenotypes_real <- jensen_bulk[1,] |> 
  pivot_longer(cols = -c(1:5), 
               names_to = "gene_treat", 
               values_to = "id") |>
  mutate(gene_treat = str_remove(gene_treat, "\\..*")) |>
  select(gene_treat, id)

```

                              
```{r prep phenos , fig.width = 14, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(stringr)

# Turn music output into graph-friendly dataframe
decon.melt <- decon.melt |>
  mutate(
    CellType = factor(CellType, levels = unique(CellType)),
    Sub = factor(Sub, levels = unique(Sub)),
    Prop = as.numeric(Prop)
  )

# Bring in real phenotype data
decon.melt <- merge(decon.melt, phenotypes_real, by.x = "Sub", by.y = "id", all.x = TRUE) |>
  mutate(
    gene_treat = factor(gene_treat, levels = unique(gene_treat))
  )

# Add genotype info
decon.melt <- decon.melt |>
  mutate(
    genotype = factor(ifelse(!is.na(gene_treat), str_split(gene_treat, " ", simplify = TRUE)[,1], NA)),
    treatment = factor(ifelse(!is.na(gene_treat), str_split(gene_treat, " ", simplify = TRUE)[,2], NA)),
    type = factor(ifelse(str_detect(Sub, "RNA|Sham"), "fraction", "whole")),
    origin = factor(case_when(
      str_detect(Sub, "B6") ~ "Rau",
      str_detect(Sub, "Sham") ~ "Froese",
      TRUE ~ "Jensen"
    )),
    fraction_type = factor(case_when(
      str_detect(Sub, "CM") ~ "Cardiomyocytes",
      str_detect(Sub, "Fib|FB") ~ "Fibroblasts",
      str_detect(Sub, "Endo|EC_") ~ "Endothelial Cells",
      TRUE ~ NA_character_
    ))
  )





```

```{r music plot sham, fig.width = 19, fig.height= 9, echo=FALSE, message=FALSE, warning=FALSE}

# Reorder the factor levels of CellType so that they are ordered by sum of their values accross groups
decon.melt$CellType <- reorder(decon.melt$CellType, -decon.melt$Prop, FUN = sum)

# Setting the levels
decon.melt$gene_treat <- factor(decon.melt$gene_treat, 
                                        levels = c('WT sham', 'WT MI', 'KO sham', 'KO MI'))

#! These colors are hard to read when printed, should be lighter
#! Or the dots from geom_jitter need to be bigger/brighter

my_palette <- c('#78c679', '#006d2c', '#d95f0e', '#993404')

plot.jen.sample <- decon.melt |>
  subset(type == "whole") |>
  ggplot(aes(x = as.numeric(CellType), y = Prop, fill = gene_treat)) +
  geom_bar(stat = "summary", 
           position = position_dodge(0.9), 
           fun = mean,
           width = 0.9,
           color = "black") +
  theme(axis.text.x = element_text(color = "black", size = 25, angle =15),
        axis.text.y = element_text(color = "black", size = 20),
        axis.ticks = element_blank(),
        legend.position = c(0.95, 0.95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6),
        strip.text = element_text(size = 20),
        title = element_text(size = 30),
        legend.text = element_text(size = 25),
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype-Treatment") +
  scale_fill_manual(values = my_palette) +  # using custom palette
  geom_jitter(inherit.aes = T, 
              position = position_jitterdodge(0.01, 0, 0.9),
              size = 2) +
  scale_x_continuous(
    breaks = 1:length(levels(decon.melt$CellType)),
    label = levels(decon.melt$CellType)) +
  geom_magnify(from = c(1.55,2.45,0,0.2), to = c(1.7,4,0.33,0.8), 
                   shape = "rect", shadow = TRUE) 

plot.jen.sample
```

```{r bcvs.fractions, fig.width = 16, fig.height= 4, echo=FALSE, message=FALSE, warning=FALSE}

# Plot the samples which are purified, bulk sequenced cell types
plot.fractions <- decon.melt |>
  subset(type == "fraction") |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 0.8,
           color = "black")+
   scale_color_brewer(palette = "dark2")  +
   facet_wrap(~fraction_type, 
           scales = "free_x",
           labeller =  label_wrap_gen(multi_line=FALSE)) +
   ylab("Proportion") +
   theme_minimal() +
   theme(axis.text.x = element_blank(),
        axis.text.y = element_text(color = "black", size = 20),
        axis.ticks = element_blank(),
        strip.text = element_text(size = 30),
        title = element_text(size = 30),
        legend.text = element_text(size = 25),
        panel.background = element_blank(),
        panel.grid = element_blank()) +
   xlab("Sample replicates") 

plot.fractions

```

```{r dirichlet, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

## Prep DirichletReg matrix

# Add small value to remove zeros
decon.melt$Prop <- decon.melt$Prop + 0.0001

# Make model matrix
dir.mat <- decon.melt |>
  subset(type == "whole",
         select = c("Sub", "CellType", "Prop", "genotype", "treatment")) |>
  dcast(Sub + genotype + treatment ~ CellType, value.var = "Prop")


### TEMP ###

############

# Convert data to DirichletRegData object
dir.mat$CellTypes <- DR_data(dir.mat[,c(4:length(dir.mat))])

dir.mat$genotype   <- as.factor(dir.mat$genotype) |>
                        relevel(ref = "WT") 

dir.mat$treatment <- as.factor(dir.mat$treatment) |>
                        relevel(ref = "sham") 
# Run Dirichlet regression
model.1 <- DirichReg(CellTypes ~ treatment * genotype, data = dir.mat, model = "alternative", base = 3)
model.2 <- DirichReg(CellTypes ~ treatment * genotype, data = dir.mat, model = "common")

# compare models, find if interaction term improves model

summary(model.1)
model.1$sub.comp

```

```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
library(makeunique)
jensen_bulk <- read_xlsx("jensen/data/raw/jensen_counts_correct.xlsx")[-1,-c(1:4,20,21)] |>
               as.data.frame()
colnames(jensen_bulk) <- c("gene", make_unique(unlist(phenotypes_real$gene_treat)))

phenotypes_real$de_id <- colnames(jensen_bulk)[-1]

# Add a suffix to duplicate gene names to make them unique
make.names.jensen_bulk <- function(jensen_bulk){
    duplicated.names <- duplicated(jensen_bulk$gene)
    suffix <- cumsum(duplicated.names)
    jensen_bulk$gene[duplicated.names] <- paste0(jensen_bulk$gene[duplicated.names], ".", suffix[duplicated.names])
    return(jensen_bulk)
}

jensen_bulk <- make.names.jensen_bulk(jensen_bulk)

# Now set the row names
row.names(jensen_bulk) <- jensen_bulk$gene

#write.csv(jensen_bulk, "jensen/results/deseq2_08032023/bulk_data")

#write.csv(dir.mat$CellTypes, "jensen/results/deseq2_08032023/comp_data")

```


```{r clr transform}
# use clr or ilr to incorp the compositions into the design matrix
comps.clr <- compositions::clr(dir.mat$CellTypes)

colnames(comps.clr) <- paste0("clr.", colnames(comps.clr))

```

```{r pca}

pca <- prcomp(t(dir.mat$CellTypes[,1:3]))

summary(pca)
```
```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

# Prepare for DESeq2
count_data <- jensen_bulk[,-1]

count_data <- mutate_all(count_data, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers

# Prepare sample information
sample_info <- data.frame(
  row.names = colnames(count_data),
  genotype =  lapply(strsplit(colnames(count_data), " "),  "[[", 1) |> unlist() |> as.factor() ,
  treatment = lapply(strsplit(colnames(count_data), " "),  "[[", 2) |> unlist() |> as.factor()
)

sample_info$genotype <- relevel(sample_info$genotype, ref = "WT")
sample_info$treatment <- relevel(sample_info$treatment, ref = "sham")

```

```{r deseq pca}

sample.pca <- cbind(sample_info, pca$rotation)

# Create a DESeqDataSet
dds.pca <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample.pca,
  design = ~ genotype + treatment + genotype:treatment + PC1 
)

# Run the DESeq pipeline
dds.pca <- DESeq(dds.pca)

# Run the differential expression analysis
res.pca <- results(dds.pca, name="genotypeKO.treatmentMI")

```

```{r deseq clr}
sample.clr <- cbind(sample_info, comps.clr)

# Create a DESeqDataSet
dds.clr <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample.clr,
  design = ~ genotype + treatment + genotype:treatment + clr.cardiomyocytes +  clr.fibroblasts #+ clr.endothelial_cells
)

# Run the DESeq pipeline
dds.clr <- DESeq(dds.clr)

# Run the differential expression analysis
res.clr <- results(dds.clr, name="genotypeKO.treatmentMI")

```

```{r deseq raw}
# Create a DESeqDataSet
dds.raw <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample_info,
  design = ~ genotype + treatment + genotype:treatment
)

# Run the DESeq pipeline
dds.raw <- DESeq(dds.raw)

# Run the differential expression analysis
res.raw <- results(dds.raw, name="genotypeKO.treatmentMI")
```


```{r compare deseq pca, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
# Compare results
comparison.pca <- merge(as.data.frame(res.pca), as.data.frame(res.raw), by = "row.names", suffixes = c(".pca", ".raw"))

# Calculate difference in -log10 p-values
comparison.pca$p_diff <- -log10(comparison.pca$padj.pca) - -log10(comparison.pca$padj.raw)

# Make categorical column
comparison.pca <- comparison.pca |>
        mutate(pca.sig = ifelse(padj.pca <= 0.05 & padj.raw >= 0.05, T, F),
               raw.sig = ifelse(padj.pca >= 0.05 & padj.raw <= 0.05, T, F),
               both.sig = ifelse(padj.pca <= 0.05 & padj.raw <= 0.05, T, F)) 

# Get the gene names for the top 5 genes by change in adjusted p value, within lost and gained sig groups
gained.pca <- comparison.pca |>  
  subset(pca.sig == T) |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

lost.pca <- comparison.pca |>  
  subset(raw.sig == T) |> 
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

top.pca <- comparison.pca |> 
  arrange(padj.pca) |>
  slice_head(n = 15)
```

```{r plot pca, fig.width = 14, fig.height= 12}
limit <- comparison.pca |> 
  na.omit() |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 1) |>
  pull(p_diff) * c(1,-1)

plot.pca <- ggplot(comparison.pca, aes(x = log2FoldChange.pca, y = -log10(padj.pca), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_distiller(type = "div", limit = limit)+
  geom_text_repel(data = gained.pca, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 40, segment.linetype = 2, segment.size = 0.6, color = "orange", force_pull = 0.01, min.segment.length = 0) + # label most increase sig genes 
  geom_text_repel(data = lost.pca, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "purple", force_pull = 0.01, min.segment.length = 0) + # label most decrease sig genes 
  geom_text_repel(data = top.pca, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) + # label most sig genes 
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
        panel.background = element_rect(fill='transparent'),
        plot.background = element_rect(fill='transparent', color=NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  ggtitle("DE w/ PCA-based composition in model", subtitle = "Colored text show genes that became or lost significance, top 5 by change")# move legend to the bottom

plot.pca
```

```{r compare deseq clr, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
# Compare results
comparison.clr <- merge(as.data.frame(res.clr), as.data.frame(res.raw), by = "row.names", suffixes = c(".clr", ".raw"))

# Calculate difference in -log10 p-values
comparison.clr$p_diff <- -log10(comparison.clr$padj.clr) - -log10(comparison.clr$padj.raw)

# Make categorical column
comparison.clr <- comparison.clr |>
        mutate(clr.sig = ifelse(padj.clr <= 0.05 & padj.raw >= 0.05, T, F),
               raw.sig = ifelse(padj.clr >= 0.05 & padj.raw <= 0.05, T, F),
               both.sig = ifelse(padj.clr <= 0.05 & padj.raw <= 0.05, T, F)) 

# Get the gene names for the top 5 genes by change in adjusted p value, within lost and gained sig groups
gained.clr <- comparison.clr |>  
  subset(clr.sig == T) |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

lost.clr <- comparison.clr |>  
  subset(raw.sig == T) |> 
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

top.clr <- comparison.clr |> 
  arrange(padj.clr) |>
  slice_head(n = 15)

top.lost.raw <- comparison.clr |>
  subset(raw.sig == T) |> 
  arrange(padj.raw) |>
  slice_head(n = 10)
```

```{r plot clr, fig.width = 14, fig.height= 12}
limit <- comparison.clr |> 
  na.omit() |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 1) |>
  pull(p_diff) * c(1, -1)

plot.clr <- ggplot(comparison.clr, aes(x = log2FoldChange.clr, y = -log10(padj.clr), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_distiller(type = "div", limit = limit)+
  geom_text_repel(data = gained.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 40, segment.linetype = 2, segment.size = 0.6, color = "orange", force_pull = 0.01, min.segment.length = 0) + # label most increase sig genes 
  geom_text_repel(data = lost.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "purple", force_pull = 0.01, min.segment.length = 0) + # label most decrease sig genes 
  geom_text_repel(data = top.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) + # label most sig genes 
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
        panel.background = element_rect(fill='transparent'),
        plot.background = element_rect(fill='transparent', color=NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  ggtitle("DE w/ clr-based composition in model", subtitle = "Colored text show genes that became or lost significance, top 5 by change")# move legend to the bottom

plot.clr
```


































```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

# Prepare for DESeq2
count_data <- jensen_bulk[,-1]

count_data <- mutate_all(count_data, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers

# Prepare sample information
sample_info <- data.frame(
  row.names = colnames(count_data),
  genotype =  lapply(strsplit(colnames(count_data), "[ ]"),  "[[", 1) |> unlist() |> as.factor() ,
  treatment = lapply(strsplit(colnames(count_data), "[.]"),  "[[", 2) |> unlist() |> as.factor()
)

sample_info$genotype <- relevel(sample_info$genotype, ref = "WT")
sample_info$treatment <- relevel(sample_info$treatment, ref = "sham")

```

```{r deseq pca}

sample.pca <- cbind(sample_info, pca$rotation)

# Create a DESeqDataSet
dds.pca <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample.pca,
  design = ~ genotype + treatment + genotype:treatment + PC1 
)

# Run the DESeq pipeline
dds.pca <- DESeq(dds.pca)

# Run the differential expression analysis
res.pca <- results(dds.pca, name="genotypeKO.treatmentMI")

```

```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

# Prepare for DESeq2
count_data <- jensen_bulk[,-1]

count_data <- mutate_all(count_data, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers

# Prepare sample information
sample_info <- data.frame(
  row.names = colnames(count_data),
  genotype =  lapply(strsplit(colnames(count_data), " "),  "[[", 1) |> unlist() |> as.factor() ,
  treatment = lapply(strsplit(colnames(count_data), " "),  "[[", 2) |> unlist() |> as.factor()
)

sample_info$genotype <- relevel(sample_info$genotype, ref = "WT")
sample_info$treatment <- relevel(sample_info$treatment, ref = "sham")

# Make sure the sample information matches the count data
all(rownames(sample_info) == colnames(count_data))

# Create a DESeqDataSet
dds <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample_info,
  design = ~ genotype + treatment + genotype:treatment  # include interaction term
)

# Run the DESeq pipeline
dds <- DESeq(dds)

resultsNames(dds)

# Run the differential expression analysis
res <- results(dds, name="genotypeKO.treatmentMI")

```

```{r deseq w comp, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
sample_info_comp <- cbind(sample_info, dir.mat[,c(4:8)])
colnames(sample_info_comp)[5] <- "endothelial_cells"
# Make sure the sample information matches the count data
all(rownames(sample_info_comp) == colnames(count_data))

# Create DESeqDataSet w/ variable compositions in design
dds_composition  <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample_info_comp,
  design = ~ genotype + treatment + genotype:treatment + cardiomyocytes + fibroblasts 
)

# Run 
dds_composition <- DESeq(dds_composition)

# Get interaction results
res_composition <- results(dds_composition, name = "genotypeKO.treatmentMI")

```

```{r compare deseq outs}
# Compare results
comparison <- merge(as.data.frame(res), as.data.frame(res_composition), by = "row.names", suffixes = c(".no_comp", ".comp"))

# Add a column indicating whether a gene is significant in either analysis
comparison$significant <- (comparison$pvalue.no_comp < 0.05) | (comparison$pvalue.comp < 0.05)

# Load ggplot2
library(ggplot2)

# Scatterplot of p-values
ggplot(comparison, aes(x = -log10(pvalue.no_comp), y = -log10(pvalue.comp), color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "-log10(p-value) without composition", y = "-log10(p-value) with composition")

# Scatterplot of log fold changes
ggplot(comparison, aes(x = log2FoldChange.no_comp, y = log2FoldChange.comp, color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "log2(Fold Change) without composition", y = "log2(Fold Change) with composition")

# Load necessary library
library(viridis)

# Calculate difference in -log10 p-values
comparison$p_diff <- -log10(comparison$pvalue.comp) - -log10(comparison$pvalue.no_comp)

# Generate volcano plot
ggplot(comparison, aes(x = log2FoldChange.comp, y = -log10(padj.comp), color = p_diff)) +
  geom_point(alpha = 0.5) +
  scale_color_viridis() +
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Difference in -log10(p-value)")+
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +  # add a line for p=0.05
  theme(legend.position = "bottom")  # move legend to the bottom


```

```{r deseq w comp graphs, fig.width = 11.5, fig.height= 10.7, echo=TRUE, message=TRUE, warning=FALSE}
# Load ggplot2
library(ggplot2)

# Scatterplot of p-values
ggplot(comparison, aes(x = -log10(pvalue.no_comp), y = -log10(pvalue.comp), color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "-log10(p-value) without composition", y = "-log10(p-value) with composition")

# Scatterplot of log fold changes
ggplot(comparison, aes(x = log2FoldChange.no_comp, y = log2FoldChange.comp, color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "log2(Fold Change) without composition", y = "log2(Fold Change) with composition")

# Load necessary library
library(viridis)

# Calculate difference in -log10 p-values
comparison$p_diff <- -log10(comparison$pvalue.comp) - -log10(comparison$pvalue.no_comp)
hist(comparison$p_diff)


comparison <- comparison|>
                mutate(up_down = ifelse(p_diff >= 0,"up","down")) 
# Load necessary library
library(ggrepel)

# Get the gene names for the top 5 genes by padj.comp
top_genes <- comparison |> 
  group_by(up_down) |>  # take the top 5 rows
  arrange(desc(abs(p_diff)), .by_group = TRUE) |>
  na.omit() |>
  slice_head(n = 5) 

ggplot(comparison, aes(x = log2FoldChange.comp, y = -log10(padj.comp), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_viridis(option = "plasma") +
  geom_text_repel(data = top_genes, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) +
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank())  # move legend to the bottom

 # most changed genes

# Sort the dataframe by p_diff in ascending order
sorted_comparison <- comparison[order(comparison$p_diff), ]

# Display the top n genes that had the largest increase in p-value
n <- 10
top_n_genes <- head(sorted_comparison, n)
print(top_n_genes)

do_DimPlot(sn, reduction = "umap", label = T, repel = T)  +  
  NoLegend() + theme(title = element_text(size = 22))

```

```{r deseq w comp graphs, fig.width = 30, fig.height= 8, echo=TRUE, message=TRUE, warning=FALSE}
plotGene <- function(seurat, slot = "scale.data", gene, plot.title = T, plot.legend = T){
  
  seurat |>
    do_FeaturePlot(gene, reduction = "umap", slot = slot, 
                   legend.position = if(plot.legend == T){"bottom"}else{"none"}, 
                   legend.width = 2, legend.length = 25,  order = T) +
    theme(title = element_text(size = 18),
          legend.title = element_blank()) +
    labs(title = if(plot.title == T){paste0(gene, "\n-log10(p-value) change: " , names(gene))}else{" "})
}

genes_up <- top_genes |>
            subset(up_down == "up") |>
            pull(Row.names)

names(genes_up) <- top_genes |>
            subset(up_down == "up") |>
            pull(p_diff) |>
            round(digits = 2)

genes_down <- top_genes |>
            subset(up_down == "down") |>
            pull(Row.names)

names(genes_down) <- top_genes |>
            subset(up_down == "down") |>
            pull(p_diff) |>
            round(digits = 2)
# plot new
plots <- c()

plots[["cell"]] <- do_DimPlot(sn, reduction = "umap", label = T, repel = T)  +  
  NoLegend() + theme(title = element_text(size = 22))

plots.up <- lapply(names(genes_up), function(x){plotGene(seurat = sn, slot = "data", gene = genes_up[x], plot.legend = F)})

plots.down <- lapply(names(genes_down), function(x){plotGene(seurat = sn, slot = "data", gene = genes_down[x], plot.legend = F)})


plots.up <- c(plots, plots.up)

plots.down <- c(plots, plots.down)

wrap_plots(plots.up, 
           ncol = 6)
wrap_plots(plots.down, 
           ncol = 6)
```