
libs <- c("tidyverse", "Seurat", "SeuratDisk", "SCpubr", "Biobase", "MuSiC", 
          "reshape2", "SingleCellExperiment") # list libraries here
lapply(libs, require, character.only = T)
rm(libs)


sn <- LoadH5Seurat("/proj/raulab/users/brian/r_projects/cc_composition/data/processed/single_cell/merges/rau_patterson/09132023/doublets_below_3.h5seurat")
#sn <- LoadH5Seurat("data/processed/single_cell/merged_old.h5seurat")


#Load just Rau, raw counts
fractions <- read.csv("/proj/raulab/users/brian/r_projects/cc_composition/data/raw/rau_frac/celltype_counts.csv", row.names = 1)
fractions.pheno <- read.csv("/proj/raulab/users/brian/r_projects/cc_composition/data/raw/rau_frac/celltype_pheno.csv")


library(makeunique)
jensen_bulk <- readxl::read_xlsx("jensen/data/raw/jensen_counts_correct.xlsx")[,-c(20,21)]|>
               as.data.frame()

colnames(jensen_bulk)[c(20,21)] <- c("WT MI", "KO MI")

phenotypes_real <- jensen_bulk[1,] |> 
  pivot_longer(cols = -c(1:5), 
               names_to = "gene_treat", 
               values_to = "id") |>
  mutate(gene_treat = str_remove(gene_treat, "\\..*")) |>
  select(gene_treat, id)

jensen_bulk <- jensen_bulk[-1,-c(1:4)]
colnames(jensen_bulk) <- c("gene", make_unique(unlist(phenotypes_real$gene_treat)))

phenotypes_real$de_id <- colnames(jensen_bulk)[-1]

# Add a suffix to duplicate gene names to make them unique
make.names.jensen_bulk <- function(jensen_bulk){
    duplicated.names <- duplicated(jensen_bulk$gene)
    suffix <- cumsum(duplicated.names)
    jensen_bulk$gene[duplicated.names] <- paste0(jensen_bulk$gene[duplicated.names], ".", suffix[duplicated.names])
    return(jensen_bulk)
}

jensen_bulk <- make.names.jensen_bulk(jensen_bulk)

# Now set the row names
row.names(jensen_bulk) <- jensen_bulk$gene

jensen_bulk <- jensen_bulk[,-1]

jensen_bulk <- mutate_all(jensen_bulk, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers
phenotypes_real$de_id <- colnames(jensen_bulk)


common.genes <- row.names(fractions)[row.names(fractions) %in% row.names(jensen_bulk)]

bulk.all <- cbind(jensen_bulk[common.genes,], fractions[common.genes,])


# Update sn with annotations
sn.mark <- sn |>
      #subset(idents = c(0,1,2,3,4,5)) |> 
      subset(features = rownames(sn)[rownames(sn) %in% rownames(bulk.all)])

sn.sce <- sn.mark |> 
  as.SingleCellExperiment(assay = "RNA") |>
  as("SummarizedExperiment")

# Add gene names
rownames(sn.sce) <- rownames(sn.mark@assays$RNA@counts)

# Get markers based on expression ratios between clusters and 1vAll comparisons

scn.markers  <- scran::findMarkers(sn.sce, groups = Idents(sn.mark), pval.type = "some", assay.type = "counts")


.getMarkers <- function(type){
marker <- scn.markers@listData[[type]] |>
  as.data.frame() %>%
  dplyr::filter(summary.logFC >= 1 |
                   summary.logFC <= -1) |>
  arrange(FDR) |>
  slice_head(n = 10) |>
  dplyr::select(p.value, FDR, summary.logFC) |>
  mutate(celltype = type) |> 
  rownames_to_column(var = "gene")
return(marker)}

# Collapse list of data frames
all.markers <- lapply(levels(Idents(sn.mark)), function(x){.getMarkers(x)}) |>
               purrr::reduce(full_join)

all.markers |>
  group_by(celltype) |>
  arrange(desc(summary.logFC)) |>
  slice_head(n = 3) 

gc()

# Genes here are printed and manuallly entered into ToppGene
# https://toppgene.cchmc.org/
genes <- function(x){ all.markers |> 
                      subset(celltype == x) |> 
                      pull(gene) |> 
                      as.data.frame() |> 
                      print(row.names = F, print.keys = F)
}

cell.types <- c("Endothelial Cells",
                "Cardiomyocytes",
                "Fibroblast",
                "VSMC/Pericytes",
                "Monocytes/MF",
                "SMC")
                #"Cardiac Neuron",
                #"?Matrix FB, Lung?")
                #"unclear",#"??EC, CA??"
                #"unclear")#"???")

sn.mark <- sn |>
 subset(idents = seq(0,length(cell.types)-1,1)) 
 #subset(idents = c(0,1,2,4,5))
names(cell.types) <- levels(sn.mark)
sn.mark <- RenameIdents(sn.mark, cell.types)

bulk.es.exp <- bulk.all[all.markers$gene,] |> 
  sapply(as.integer) %>%
  ExpressionSet(assayData = .) |> 
  exprs()
row.names(bulk.es.exp) <- all.markers$gene


# Run MuSiC with subset genes

# Convert to SingleCellExperiment
sce <- as.SingleCellExperiment(sn.mark, assay = "RNA")
  
# Exclude specified clusters
cells <- levels(Idents(sn.mark))
cells <- cells[!(cells %in% c("unclear")) & !is.na(cells)]
  
# Use MuSiC to estimate cell type proportions
decon <- music_prop(bulk.mtx = bulk.es.exp, sc.sce = sce, markers = all.markers$gene,
                      clusters = "ident", samples = "orig.ident",
                      select.ct = cells)
  
# Turn MuSiC output into graph-friendly dataframe
decon.melt <- reshape2::melt(decon$Est.prop.weighted)
colnames(decon.melt) = c('Sub', 'CellType', 'Prop')

rm(decon)
rm(sce)
rm(cells)

#write.csv(decon.melt, "results/composition_est/06222022/rau_patterson/10142023.csv", 
#          row.names = F)

```

```{r reformat data}
# Divide fraction and whole data
#decon.melt$version <- "music" 
#decon.melt <- rbind(decon.melt, props.bisque)
colnames(fractions.pheno)[4] <- "Sub"
decon.frac <- decon.melt|> 
              subset(Sub %in% fractions.pheno$Sub) |> 
              merge(fractions.pheno)


decon.whole <- decon.melt|> 
              #mutate(Sub = paste0("S", Sub)) |> 
              subset(Sub %in% phenotypes_real$de_id) |> 
              mutate(Genotype = factor(case_when(
    str_detect(Sub, "WT") ~ "WT",
    str_detect(Sub, "KO") ~ "KO")),
                     Treatment = factor(case_when(
    str_detect(Sub, "sham") ~ "Sham",
    str_detect(Sub, "MI") ~ "TAC")))

# Unmelt for clustering 
decon.wide <- decon.whole  |> 
              select(Sub, CellType, Prop) |> 
              pivot_wider(names_from = "Sub", values_from = "Prop") |> 
              column_to_rownames("CellType") %>%
              mutate_all(as.numeric)

# Reorder the factor levels of CellType so that they are ordered by sum of their values across groups
decon.melt$CellType <- reorder(decon.melt$CellType, -decon.melt$Prop, FUN = sum)

```

```{r plot fractions, fig.width=14, fig.height = 5}
 decon.frac <- decon.frac |> 
  mutate(cell.type = factor(case_when(
    str_detect(cell.type, "CM") ~ "Cardiomyocytes",
    str_detect(cell.type, "Endo") ~ "Endothelial Cells",
    str_detect(cell.type, "Fib") ~ "Fibroblasts")))

decon.frac |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 1,
           color = "black")+
  scale_fill_brewer(name = "Cell Type",
                      palette = "Set2")  +
   facet_wrap(~cell.type,
              scales = "free_x")+
   ylab("Proportion") +
   theme_minimal() +
   theme(strip.text = element_text(size = 30),
          title = element_text(size = 20),
          axis.text.y  = element_text(size = 20),
          legend.text = element_text(size = 22),
          axis.ticks.x = element_blank(),
          panel.grid = element_blank(),
          legend.title = element_blank(),
          legend.position = 'right',
          legend.justification = 'center',
          legend.margin = margin(t = -5, r = 0, b = 0, l = 0, unit = "pt"),
          axis.text.x = element_blank(),
          plot.margin = margin(0, 5, 1, 5),
          plot.caption = element_text(size = 22, hjust = -0.3, face = "bold"),
          plot.tag = element_text(size = 22, hjust = -0.3),
          axis.title.y = element_text(size = 25),
          axis.title.x =  element_text(size = 25)) +
   xlab("Pure Cell Type Bulk RNAseq Replicates") 


```

```{r comp plots, echo=FALSE, fig.width=15, fig.height = 7}


library(RColorBrewer)
brewer.pal(n=8,"Paired")

my_palette <- c("#A6CEE3", "#1F78B4", "#FDBF6F", "#FF7F00")
legend.names <- c("Sham_1","Sham_2", "TAC_1", "TAC_2")


ordered_df <- decon.whole %>%
  filter(CellType == "Cardiomyocytes") %>%  # Filter rows where CellType is "Cardiomyocytes"
  arrange(Prop) %>%  # Sort these rows by Prop in descending order
  pull(Sub)  # Extract the Sub values in this order
# Use the ordered Sub values to reorder the original data frame
decon.whole <- decon.whole %>%
  mutate(Sub = factor(Sub, levels = ordered_df)) %>%
  arrange(Sub)
decon.whole$CellType_wrap = str_wrap(decon.whole$CellType, width = 12)

  
# Use the ordered Sub values to reorder the original data frame
decon.whole <- decon.whole %>%
  mutate(Sub = factor(Sub, levels = ordered_df)) %>%
  arrange(Sub) |> 
  mutate(Genotype_Treatment = paste(Genotype, " - ", Treatment))


cell.type.order <-  decon.whole %>%
  group_by(CellType_wrap) |> 
  mutate(mean = mean(Prop)) |> 
  arrange(desc(mean)) |>
  pull(CellType_wrap) |> 
  unique()

design <- "AA
           BB"

# Generate boxplot
comp_celltype <- decon.whole   %>%
    ggplot(aes(x = factor(CellType_wrap, levels = as.character(cell.type.order)), y = Prop, fill = Genotype_Treatment)) +
   #geom_boxplot(position = position_dodge(0.9), width = 0.9, color = "black") +
   geom_bar(stat = "summary",position = position_dodge(0.9),  fun = mean,width = 0.9,  color = "black", alpha = 0.8) +
    geom_jitter(inherit.aes = T, 
                position = position_jitterdodge(jitter.width = 0.005, dodge.width = 0.9),
                size = 4, alpha = 0.5) +
    theme(axis.text.x = element_text(color = "black", size = 28, angle =25, vjust = 0.5),
          axis.text.y = element_text(color = "black", size = 28),
          axis.ticks = element_blank(),
          legend.position = c(0.9, 1),
          legend.justification = c("right", "top"),
          legend.box.just = "right",
          legend.margin = margin(6, 6, 6, 6),
          strip.text = element_text(size = 20),
          title = element_text(size = 30),
          legend.text = element_text(size = 32),
          panel.background = element_rect(fill='transparent'),
          plot.background = element_rect(fill='transparent', color=NA),
          #panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          legend.background = element_rect(fill='transparent'),
          legend.box.background = element_rect(fill='transparent'),
          axis.title.x = element_blank(),
          plot.margin = unit(c(1,1,1,1), units = "cm"),
          axis.title.y = element_text(size = 30)) +
    labs(y = "Proportion", 
         fill = "Treatment") +
    scale_fill_manual(values = my_palette)
  
comp_celltype

#patchwork::wrap_plots(A = comp_celltype, B = comp_sample, design = design)


```

```{r change relative graph prep, fig.width=15, fig.height = 7}

control.means <- decon.whole |> group_by(CellType, Genotype) |> 
            subset(Treatment == "Sham") |> 
            summarize(cont.mean = mean(Prop)) %>%
            left_join(decon.whole, .) |>
            mutate(props.norm = Prop/cont.mean) 
            

# Generate boxplot
p.box <- control.means   %>%
  subset(Treatment == "TAC") |> 
    ggplot(aes(x = factor(CellType_wrap, levels = as.character(cell.type.order)), y = props.norm, fill = Genotype)) +
    geom_boxplot(position = position_dodge(0.8), width = 0.8, color = "black") +
    theme(axis.text.x = element_text(color = "black", size = 28, angle =25, vjust = 0.5),
          axis.text.y = element_text(color = "black", size = 28),
          legend.position = c(0.9,0.9),
          legend.margin = margin(6, 6, 6, 6),
          strip.text = element_blank(),
          title = element_text(size = 30),
          legend.text = element_text(size = 32),
          legend.title = element_text(size = 32),
          panel.background = element_rect(fill='white'),
          panel.grid.major = element_line(color = "darkgrey",
                                          size = 0.5),
          panel.grid.minor = element_blank(),
          legend.box.background = element_rect(fill='transparent'),
          axis.title.x = element_blank(),
          plot.margin = unit(c(1,2,1,1), units = "cm"),
          axis.title.y = element_text(size = 30),
          axis.text.x.top = ) +
    labs(y = "Proportion change\nrelative to control", 
         fill = "Treatment") +
    scale_fill_manual(values = my_palette) 

p.box

```

```{r dirichlet, fig.width=15, fig.height = 7}

dir.model <- decon.whole |> 
  mutate(CellType = factor(CellType),
         Genotype = factor(Genotype),
         Treatment = factor(Treatment)
  ) 

## Prep DirichletReg matrix

# Add small value to remove zeros
dir.model$Prop <- dir.model$Prop + 0.0001

# Make model matrix
dir.mat <- dir.model |>
  subset(select = c("Sub", "CellType", "Prop", "Treatment", "Genotype")) |>
  dcast(Sub + Treatment + Genotype ~ CellType, value.var = "Prop")


library(DirichletReg)
# Convert data to DirichletRegData object
dir.mat$CellTypes <- DR_data(dir.mat[,c(4:length(dir.mat))])

dir.mat$Treatment <- as.factor(dir.mat$Treatment) |>
                        relevel(ref = "Sham") 

dir.mat$Genotype <- as.factor(dir.mat$Genotype) |>
                        relevel(ref = "WT")
# Run Dirichlet regression
#model.1 <- DirichReg(CellTypes ~ Treatment, data = dir.mat, model = "alternative", base = 3)
model.2 <- DirichReg(CellTypes ~ Treatment * Genotype, data = dir.mat, model = "common")

# compare models, find if interaction term improves model

dir.results <- summary(model.2)[["coef.mat"]] |> as.data.frame()
class(model.2)
model.2$logLik
```

```{r, fig.width=18, fig.height=8}
# You left off here on 12/6/23
# dir.results is a matrix of the model output 
# Needs to be reformatted to make error plot
# Now, plot the data
dir.results$CellType <- rep(summary(model.2)[["varnames"]], each = 4)
dir.results$Feature <- row.names(summary(model.2)[["coef.mat"]])
colnames(dir.results)[2] <- "StdError"
dir.results <- dir.results |> 
  mutate(Feature_wrap = case_when(
    str_detect(Feature, "TreatmentTAC$") ~ "Myocardial\nInfarction",
    str_detect(Feature, ":") ~ "Myocardial\nInfarction\nand KO",
    str_detect(Feature, "GenotypeKO$") ~ "A1-AR KO"
  ))

err.plot <- dir.results |> 
  subset(Feature != "(Intercept)" & `Pr(>|z|)` < 0.05) |> 
    ggplot(aes(y = Feature_wrap, x = Estimate, color = CellType)) +
    geom_point(position = position_dodge(width = 0.6), size = 6) +
    geom_errorbarh(aes(xmin = Estimate - StdError, xmax = Estimate + StdError),
                  height = 0.2, position = position_dodge(width = 0.6), size = 2) +
   # geom_text(aes(label = Significance, x = Estimate + StdError, group = CellType), 
    #          position = position_dodge(width = 0.6), 
    #          colour = "black", hjust = -0.2, size = 10) +
    scale_color_brewer(name = "Cell Type",
                      palette = "Set2")  +
    scale_x_continuous(limits = c(-5, 3), n.breaks = 8) +
    theme_classic() +
    theme(axis.text.x = element_text(color = "black", size = 20),
        axis.text.y = element_text(color = "black", size = 20, angle = 0),
        axis.ticks = element_blank(),
        legend.position = "left",
        legend.justification = c("right", "center"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6),
        title = element_text(size = 20),
        legend.text = element_text(size = 20),
        axis.title.y = element_text(color = "black", size = 24),
        axis.title.x = element_text(color = "black", size = 24),
         #panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_line(colour = "grey", size = 0.5),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')) +
  labs(x = "Coeffiecients (estimates of effect)", 
       y = "Predictor Variables", 
       legend = "Cell Types") 

err.plot
```

```{r clr transform}
# use clr or ilr to incorp the compositions into the design matrix
comps.clr <- compositions::clr(t(decon.wide))

colnames(comps.clr) <- paste0("clr.", colnames(comps.clr))
```

```{r pca transform}
pca <- prcomp(decon.wide)
```



```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

# Prepare for DESeq2

bulk <- mutate_all(jensen_bulk, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers

pheno.reorder <- phenotypes_real |> 
  mutate(Sub = as.factor(de_id)) |> 
  mutate(Genotype = factor(case_when(
    str_detect(Sub, "WT") ~ "WT",
    str_detect(Sub, "KO") ~ "KO")),
                     Treatment = factor(case_when(
    str_detect(Sub, "sham") ~ "Sham",
    str_detect(Sub, "MI") ~ "TAC")))
# Prepare sample information
sample_info <- data.frame(
  row.names = colnames(bulk),
  genotype = as.factor(pheno.reorder$Genotype),
  treatment = as.factor(pheno.reorder$Treatment)
)

sample_info$treatment <- relevel(sample_info$treatment, ref = "Sham")
sample_info$genotype <- relevel(sample_info$genotype, ref = "KO")

```

```{r deseq pca}

sample.pca <- cbind(sample_info, pca$rotation)

# Create a DESeqDataSet
dds.pca <- DESeqDataSetFromMatrix(
  countData = bulk,
  colData = sample.pca,
  design = ~ treatment + genotype + treatment:genotype + PC1
)

# Run the DESeq pipeline
dds.pca <- DESeq(dds.pca)

# Run the differential expression analysis
res.pca <- results(dds.pca, name="treatmentTAC.genotypeWT")
```


```{r deseq clr}
library(DESeq2)
sample.clr <- cbind(sample_info, comps.clr)


# Create a DESeqDataSet
dds.clr <- DESeqDataSetFromMatrix(
  countData = bulk,
  colData = sample.clr,
  design = ~ treatment + genotype + treatment:genotype + clr.Cardiomyocytes + clr.Fibroblast
)

# Run the DESeq pipeline
dds.clr <- DESeq(dds.clr)

# Run the differential expression analysis
#res.clr <- results(dds.clr, contrast = c("treatment", "Sham", "TAC"))
res.clr <- results(dds.clr, name="treatmentTAC.genotypeWT")

```

```{r deseq raw}
# Create a DESeqDataSet
dds.raw <- DESeqDataSetFromMatrix(
  countData = bulk,
  colData = sample_info,
  design = ~ treatment + genotype + treatment:genotype
)

# Run the DESeq pipeline
dds.raw <- DESeq(dds.raw)

# Run the differential expression analysis
res.raw <- results(dds.raw, name="treatmentTAC.genotypeWT")
results(dds.raw)
```

```{r compare deseq pca, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
# Compare results
comparison.pca <- merge(as.data.frame(res.pca), as.data.frame(res.raw), by = "row.names", suffixes = c(".pca", ".raw"))

# Calculate difference in -log10 p-values
comparison.pca$p_diff <- -log10(comparison.pca$padj.pca) - -log10(comparison.pca$padj.raw)

# Make categorical column
comparison.pca <- comparison.pca |>
        mutate(pca.sig = ifelse(padj.pca <= 0.05 & padj.raw >= 0.05, T, F),
               raw.sig = ifelse(padj.pca >= 0.05 & padj.raw <= 0.05, T, F),
               both.sig = ifelse(padj.pca <= 0.05 & padj.raw <= 0.05, T, F)) 

# Get the gene names for the top 5 genes by change in adjusted p value, within lost and gained sig groups
gained.pca <- comparison.pca |>  
  subset(pca.sig == T) |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

lost.pca <- comparison.pca |>  
  subset(raw.sig == T) |> 
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

top.pca <- comparison.pca |> 
  arrange(padj.pca) |>
  slice_head(n = 5)
```

```{r plot pca, fig.width = 18, fig.height= 10}
library(ggrepel)
library("wesanderson")
pal <- wes_palette(name = "Zissou1", type = "continuous")
limit <- comparison.pca |> 
  na.omit() |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 1) |>
  pull(p_diff) * c(1, -1)

plot.pca <- ggplot(comparison.pca, aes(x = log2FoldChange.raw, y = -log10(padj.raw), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_distiller(type = "div", palette = "Spectral" ) +
  #geom_text_repel(data = gained.pca, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
  #                force = 40, segment.linetype = 2, segment.size = 0.6, color = "#d95f0e", force_pull = 0.01, min.segment.length = 0) + # label most increase sig genes 
 # geom_text_repel(data = lost.pca , aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
  #                force = 20, segment.linetype = 2, segment.size = 0.6, color = "purple", force_pull = 0.01, min.segment.length = 0) + # label most decrease sig genes 
  geom_text_repel(data = top.pca |>  subset(Row.names != "Sccpdh"), aes(label = Row.names), size = 12, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) + # label most sig genes 
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.5, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
        panel.background = element_rect(fill='transparent'),
        plot.background = element_rect(fill='transparent', color=NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        plot.margin = unit(c(1,1,1,1), "cm")) #+
  #ggtitle("Heart Failure DE w/ PCA-based composition in model")#, #subtitle = "Colored text show genes with greatest sig. gain")# move legend to the bottom

plot.pca
```



```{r compare deseq clr, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
# Compare results
comparison.clr <- merge(as.data.frame(res.clr), as.data.frame(res.raw), by = "row.names", suffixes = c(".clr", ".raw"))

# Calculate difference in -log10 p-values
comparison.clr$p_diff <- -log10(comparison.clr$padj.clr) -  # low numbers mean that its not signficant when clr is used
                         -log10(comparison.clr$padj.raw)    # large numbers mean that its significant w/o comp

# Make categorical column
comparison.clr <- comparison.clr |>
        mutate(clr.sig = ifelse(padj.clr <= 0.05 & padj.raw >= 0.05, T, F),
               raw.sig = ifelse(padj.clr >= 0.05 & padj.raw <= 0.05, T, F),
               both.sig = ifelse(padj.clr <= 0.05 & padj.raw <= 0.05, T, F)) 

# Get the gene names for the top 5 genes by change in adjusted p value, within lost and gained sig groups
gained.clr <- comparison.clr |>  
  subset(clr.sig == T) |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

lost.clr <- comparison.clr |>  
  subset(raw.sig == T) |> 
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

top.clr <- comparison.clr |> 
  arrange(padj.clr) |>
  slice_head(n = 5)

top.lost.raw <- comparison.clr |>
  subset(raw.sig == T) |> 
  arrange(padj.raw) |>
  slice_head(n = 5)
```

```{r plot clr, fig.width = 11, fig.height= 12}
library(ggrepel)
library("wesanderson")
pal <- wes_palette(name = "Zissou1", type = "continuous")
limit <- comparison.clr |> 
  na.omit() |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 1) |>
  pull(p_diff) * c(-1, 1)

plot.clr <- ggplot(comparison.clr, aes(x = log2FoldChange.raw, y = -log10(padj.raw), color = p_diff)) +
  geom_point(alpha = 1, size = 6) +
  scale_color_distiller(type = "div", palette = "Spectral" ) +
  #geom_text_repel(data = gained.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
  #                force = 40, segment.linetype = 2, segment.size = 0.6, color = "orange", force_pull = 0.01, min.segment.length = 0) + # label most increase sig genes 
  #geom_text_repel(data = lost.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
  #                force = 20, segment.linetype = 2, segment.size = 0.6, color = "purple", force_pull = 0.01, min.segment.length = 0) + # label most decrease sig genes 
  geom_text_repel(data = top.clr, aes(label = Row.names), size = 12, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) + # label most sig genes 
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(size = 28, vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 28),
        axis.title.x = element_text(color = "black", size = 28),
        panel.background = element_rect(color="black"),
        plot.background = element_rect(color="black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        plot.margin = unit(c(1,1,1,1), "cm")) #+
  #ggtitle("DE w/ clr-based composition in model", subtitle = "Colored text show genes that became or lost significance, top 5 by change")# move legend to the bottom

plot.clr
```

```{r TOAST}
# The following initializes usage of Bioc devel
BiocManager::install(version='3.16')

BiocManager::install("TOAST")
library("TOAST")
vignette("TOAST")
head(decon.wide)
rownames(decon.wide)[c(1,4,5)] <- c("Endothelial_cells", "VSMC_pericytes", "Monocytes_MF")
pheno.reorder <- as.data.frame(pheno.reorder)
rownames(pheno.reorder) <- pheno.reorder[,4]
pheno.reorder[,5] <- as.factor(pheno.reorder[,5])
pheno.reorder[,6] <- as.factor(pheno.reorder[,6])

pheno.toast <- pheno.reorder[,-c(1,2,3)]
decon.toast <- t(decon.wide)
bulk.toast <- as.matrix(bulk)

# Reorder decon.toast and bulk.toast based on pheno.toast
ordered_samples <- rownames(pheno.toast)
decon.toast <- decon.toast[ordered_samples, ]
bulk.toast <- bulk.toast[, ordered_samples]

Design_out <- makeDesign(pheno.toast, decon.toast)

fitted_model <- fitModel(Design_out, bulk.toast)


fitted_model <- fitModel(Design_out, as.matrix(bulk))

fitted_model$all_coefs # list all phenotype names
fitted_model$all_cell_types # list all cell type names

# Initialize an empty list to store results for each cell type
results_list <- list()

# Loop through each cell type to perform differential expression analysis
for(cell_type in fitted_model$all_cell_types) {
  res_table <- csTest(fitted_model, coef = "GenotypeWT", cell_type = cell_type)
  results_list[[cell_type]] <- res_table %>% 
    filter(adj.P.Val < 0.05) %>% 
    arrange(adj.P.Val)
}

# Combine results into a single data frame
final_results <- bind_rows(results_list, .id = "CellType")


N = 300 # simulation a dataset with 300 samples
K = 3 # 3 cell types
P <- 500 # 500 features

### simulate proportion matrix
Prop = matrix(runif(N*K, 10,60), ncol=K)
Prop = sweep(Prop, 1, rowSums(Prop), FUN="/")
colnames(Prop) = c("Neuron", "Astrocyte", "Microglia")
Y <- matrix(rnorm(N*P, N, P), ncol = N)

### simulate phenotype names
design <- data.frame(disease=factor(sample(0:1,
                     size = N,replace=TRUE)),
                     age=round(runif(N, 30,50)),
                     race=factor(sample(1:3, size = N,replace=TRUE)))
Design_out <- makeDesign(design, Prop)

### fit model
fitted_model <- fitModel(Design_out, Y)



```