---
title: "AHA BCVS GTEx"
author: "Brian Gural"
date: "7/3/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Project Goals\
Estimate cellular composition in human hearts from bulk RNAseq



```{r load libs, message=FALSE, warning=FALSE, cache=F, include=F}
# load libraries
libs <- c("Seurat", "ggplot2", "DESeq2", "patchwork","SeuratDisk", "MuSiC", "reshape2",
          "tidyverse", "SingleCellExperiment","harmony", "SCpubr","shiny", 
          "AUCell", "viridis", "gplots", "scales", "ggrepel", "gridExtra", "scCustomize",
          "httr","readxl","matrixStats", "TabulaMurisSenisData", "ggforce") # list libraries here
lapply(libs, require, character.only = T)
source("gtex/scripts/functions/decon_all.R")
```


## Analysis\

```{r load data, message=FALSE, warning=FALSE, cache=TRUE, include=TRUE}
# bulk
bulk <- read.csv("gtex/data/processed/internal/gtex_lv_counts_summed.csv", check.names = F, row.names = 2)[,-1]

# Single cell of all datasets combined
sn.gtex <- LoadH5Seurat("gtex/data/processed/internal/sn_gtex_lv_match.h5seurat")
sn.gtex$origin <- "gtex"
sn.sim <- LoadH5Seurat("gtex/data/raw/external/single_cell/simonson_2023/simonson2023_raw.h5seurat")
sn.sim$origin <- "simonson"

sn.all <- merge(sn.gtex, sn.sim)
```

```{r cluster, echo=FALSE,warning=FALSE, cache=TRUE, include=TRUE}
# seurat after 1_1_jensen_doublets.R and 1_2_jensen_merge.R
# should have no doublets and less ambient RNA
# also datasets were filtered before combining for mt percent, features, and counts,
sn.clust <- sn.all |>
    FilterByQuantile() |>
    ClusterSeurat(subset = F, res = 0.05, regress.by = c("origin"))

```


### Features of GTEx snRNAseq Reference

```{r cluster dimplots, echo=FALSE,warning=FALSE, include=TRUE, fig.width= 12, fig.height=18}
plots <- c()
# seurat after 1_1_jensen_doublets.R and 1_2_jensen_merge.R
plots[["dim.clust"]] <- do_DimPlot(sn.clust, label = T, repel = T) + NoLegend()
plots[["dim.origin"]] <- do_DimPlot(sn.clust, group.by = "Participant.ID", label = T, repel = T) + NoLegend()

# Convert your table to a data frame and reset row names
df <- as.data.frame(table(sn.clust$Participant.ID, sn.clust$seurat_clusters))
colnames(df) <- c("Participant.ID", "cluster", "count")

# Convert data frame from wide format to long format

# Convert cluster to factor for ordered x-axis
df$cluster <- factor(df$cluster, levels = unique(df$cluster))

# Create the bar plot
plots[["bar.clust.orig"]] <- ggplot(df, aes(x = cluster, y = count, fill = Participant.ID)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(x = "Cluster", y = "Count", fill = "Participant.ID", 
       title = "Count of Participant.ID by Cluster") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

plots[["bar.umi"]] <- VlnPlot(sn.clust, features = "nCount_RNA", log = T) 

plots[["feat.count"]] <- do_FeaturePlot(sn.clust, features = "nCount_RNA") 
plots[["feat.features"]] <- do_FeaturePlot(sn.clust, features = "nFeature_RNA") 
plots[["feat.mito"]] <- do_FeaturePlot(sn.clust, features = "PercentMito") 
plots[["feat.doub"]] <- do_FeaturePlot(sn.clust, features = "scrublet_score") 


wrap_plots(plots, ncol = 2)
```

### How well do original GTEx annotations correlate with our current clusters? \

```{r seurat vs gtex labels heatmap, echo=FALSE,warning=FALSE, include=TRUE, fig.height = 8, fig.width = 12}
# Create a contingency table
contingency_table <- table(c(sn.clust$Broad.cell.type),
                           c(sn.clust$seurat_clusters))

# Perform the chi-squared test
chi_squared_test <- chisq.test(contingency_table)

# Calculate the standardized residuals
observed_frequencies <- contingency_table
expected_frequencies <- chi_squared_test$expected
standardized_residuals <- (observed_frequencies - expected_frequencies) / sqrt(expected_frequencies)


col_palette <- viridis(25)
# Create a heatmap of the standardized residuals
heatmap.2(standardized_residuals, 
          trace = "none", 
          col = col_palette,
          main = "Residuals, All Cell Types",
          xlab = "Seurat Clusters",
          ylab = "All Cell Types",
          margins = c(8, 12),
          key.title = "Standardized Residuals",
          srtCol = 0,
          cexRow = 1.2)

# Find the index of the largest value in each column of the normalized_contingency_table
max.indices <- apply(standardized_residuals, 2, which.max)
# Get the corresponding BroadCellType for each Seurat cluster
assigned.cell.types <- rownames(standardized_residuals)[max.indices] |>
                          make.unique()

#rename the clusters
names(assigned.cell.types) <- levels(sn.clust)
sn.labeled <- RenameIdents(sn.clust, assigned.cell.types)

```

### Do our cell type clusters express canonical markers?



```{r Cluster Assignments, echo=FALSE,warning=FALSE, include=TRUE, fig.height = 12, fig.width = 18}
plotGene <- function(seurat, slot = "scale.data", gene, plot.title = T, plot.legend = T){
 
 seurat |>
  do_FeaturePlot(gene, reduction = "umap", slot = slot, 
                 legend.position = if(plot.legend == T){"bottom"}else{"none"}, 
                 legend.width = 2, legend.length = 25,  order = T) +
  theme(title = element_text(size = 20),
        legend.title = element_blank()) +
  labs(title = if(plot.title == T){paste0(gene, ", " , names(gene))}else{" "})
}

genes <- c("PECAM1", "COL1A1", "TNNT2", "CD86")
names(genes) <- c("Endothelial Cells", "Fibroblasts", "Cardiomyocytes", "Macrophages/B Cells")


# make standard palette 
pal <- brewer_pal(type = "qual", 2)(length(levels(sn.labeled)))
names(pal) <- levels(sn.labeled)
# plot new
plots <- c()
plots[["cell"]] <- do_DimPlot(sn.labeled, reduction = "umap", label = T, repel = T, colors.use = pal, plot.title = "Flexible QC, no doublets")  +  
                                    NoLegend() + theme(title = element_text(size = 20))
plots.exp <- lapply(names(genes), function(x){plotGene(seurat = sn.labeled, slot = "data", gene = genes[x], plot.legend = F)})
            
plots.new <- c(plots, plots.exp)

wrap_plots(plots.new, 
           ncol = 3)

#SaveH5Seurat(sn.labeled, "gtex/data/processed/external/single_cell/sim_gtex", overwrite = T)
```

### Manually find markers via DeconvoBuddies

```{r DeconvoBuddies Find markers, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# find markers
library(DeconvoBuddies)
library(BisqueRNA)

# Subset # of cells to speed up script
n_cells <- 15000
cells <- sample(colnames(sn.labeled), n_cells, replace = FALSE)

# Subset sn.labeled for genes common to bulk RNAseq and cells of interest
seurat <- sn.labeled |>
    subset(cells = cells, features = rownames(sn.labeled)[rownames(sn.labeled) %in% rownames(bulk)])

# Function needs a SummarizedExperiment format
sn.sce <- as.SingleCellExperiment(seurat, assay = "RNA") |>
  as("SummarizedExperiment")

# add gene names, not sure why they broke
gene_names <- rownames(seurat@assays$RNA@counts)
rownames(sn.sce) <- gene_names

# Run marker selection pipeline from DeconvolBuddies
ratios <- get_mean_ratio2(sn.sce, cellType_col = "ident")
fc <- findMarkers_1vAll(sn.sce, cellType_col = "ident", mod = "~Participant.ID")
marker_stats <- left_join(ratios, fc, by = c("gene", "cellType.target"))

```

```{r bisque huuki, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}

# Add a rank score per cell type
marker_stats2 <- marker_stats |>
    group_by(cellType.target) |>
    mutate(rank_ratio = row_number())

# Subset for markers in top 50 rank ratio, get top 15 by logFC
marker_genes <- marker_stats2 |>
                filter(rank_ratio <= 25) |>
                group_by(cellType.target) |>
                arrange(-std.logFC, .by_group = T)|>
                slice_head(n = 15) |>
                pull(gene)


marker_genes <- marker_stats2 |>
                filter(ratio >= 2 & std.logFC >= 1.2) |>
                pull(gene)
# Select top 75 by rank per cell type for reference
top_markers <- marker_stats2 |>
                filter(rank_ratio <= 75) |> 
                group_by(cellType.target) |>
                arrange(-std.logFC, .by_group = T)



```

## Deconvolution 

### Bisque

```{r bisque, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}


# Remake seurat as sn.sce without any NA in the idents
sn.sce <- sn.labeled |>
          subset(idents = unique(Idents(sn.labeled))[!is.na(unique(Idents(sn.labeled)))]) |>
          as.SingleCellExperiment(assay = "RNA") 

# Create expression set 
exp_set_sce <- ExpressionSet(
    assayData = as.matrix(assays(sn.sce)$counts),
    phenoData = AnnotatedDataFrame(
        as.data.frame(colData(sn.sce))[c("ident", "orig.ident", "Participant.ID")]
    )
)

# Check for nuclei w/ zero expression in marker genes
exp_set_sce <- exp_set_sce[marker_genes, ]
zero_cell_filter <- colSums(exprs(exp_set_sce)) != 0
message("Exclude ", sum(!zero_cell_filter), " cells")
exp_set_sce <- exp_set_sce[, zero_cell_filter]

# Make bulk RNAseq into expression set
# exclude low expressed genes 

bulk.qc <- bulk[rowSums(bulk) > 5, ]
bulk.es <- ExpressionSet(assayData = as.matrix(bulk.qc[marker_genes,]))

# Run Bisque
est_prop <- ReferenceBasedDecomposition(
    bulk.eset = bulk.es,
    sc.eset = exp_set_sce,
    cell.types = "ident",
    subject.names = "Participant.ID",
    use.overlap = FALSE
)

est_prop$bulk.props <- t(est_prop$bulk.props)

props.bisque <- est_prop$Est.prop.long <- est_prop$bulk.props %>%
    as.data.frame() %>%
    tibble::rownames_to_column("Sample") %>%
    pivot_longer(!Sample, names_to = "cell_type", values_to = "prop")

# Format columns and add meta data for decon method
props.bisque$version <- "bisque"
colnames(props.bisque) <- c('Sub', 'CellType', 'Prop', "version")
```

### MuSiC \

```{r MuSiC marker genes, echo=FALSE, warning=FALSE, include=T}
bulk.es <- exprs(bulk.es)
props.music <- EstimateCellTypeProportions(sn.labeled, bulk.es, for.aitchison = T, sn.individuals = "Participant.ID", marker = marker_genes, cells_exclude = "Myocyte (cardiac, cytoplasmic)")$Est.prop.weighted |>
                  melt()
colnames(props.music) <- c('Sub', 'CellType', 'Prop')
props.music$version <- "music"
```

```{r MuSiC all genes, echo=FALSE, warning=FALSE, include=T}

bulk.es <- ExpressionSet(assayData = as.matrix(bulk.qc))
bulk.es <- exprs(bulk.es)
props.music.all <- EstimateCellTypeProportions(sn.labeled, bulk.es, for.aitchison = T, sn.individuals = "Participant.ID", marker = rownames(bulk.qc), cells_exclude = "Myocyte (cardiac, cytoplasmic)")$Est.prop.weighted |>
                  melt()
colnames(props.music.all) <- c('Sub', 'CellType', 'Prop')
props.music.all$version <- "music.all"
```

```{r plot aitchison , echo=FALSE, warning=FALSE}
# Format deconvolution df
decon.melt <- rbind(props.music, props.music.all)

# Type formatting
decon.melt$CellType <- factor(decon.melt$CellType, levels = unique(decon.melt$CellType))
decon.melt$Prop <- as.numeric(decon.melt$Prop)
decon.melt$version <- factor(decon.melt$version, levels = unique(decon.melt$version))

#write.csv(props.bisque[,c(1:3)],"gtex/results/estimates/bcvs/bisque_07112023")
```

### Composition in whole panel, by method

```{r plot gtex all, fig.width = 12, fig.height= 10, echo=FALSE, message=FALSE, warning=FALSE}
library(forcats)
# Find the CellType with the largest average Prop
getRank <- function(versions){
avg_prop <- decon.melt |>
  subset(version == versions) |>
  group_by(CellType) |>
  summarise(avg_prop = mean(Prop, na.rm = TRUE))
ref.cell <- avg_prop$CellType[[which.max(avg_prop$avg_prop)]]
decon.ref <- decon.melt |>
  subset(version == versions & CellType == ref.cell)
rank <- decon.ref$Sub[order(decon.ref$Prop)]
}

rank.music <- getRank("music")
rank.music.all <- getRank("music.all")

# Plot composition of every bulk sample, per decon method
# 
plotGtexAll <- function(versions, ranks =  rank.music){
  decon.melt |>
   subset(version == versions) |>
   ggplot(aes(x = factor(Sub, levels = ranks), y=Prop, fill=CellType))  +
    geom_bar(stat='identity', 
           position = "fill", 
           width = 1)+ 
    scale_fill_manual(name = "Cell Type",
                     values = pal)+
    facet_wrap(~version, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE))+ 
    ylab("Proportion") + 
    theme(
      axis.text.x = element_blank(),
      strip.text = element_blank(),
      title = element_text(size = 20),
      legend.text = element_text(size = 18),
      axis.ticks.x = element_blank()) +
    xlab("GTEx Individuals") +
    ggtitle(versions)
}

plots.all <- c()
plots.all[["music"]] <- plotGtexAll("music", rank.music)
plots.all[["music.all"]] <- plotGtexAll("music.all", rank.music.all)

wrap_plots(plots.all, ncol = 1)
```


```{r gtex pheno, fig.width = 12, fig.height= 6, echo=FALSE, message=FALSE, warning=FALSE}
pheno <- read.table(gzfile("gtex/data/raw/phenotypes/phs000424.v9.pht002742.v9.p2.c1.GTEx_Subject_Phenotypes.GRU.txt.gz"),sep="\t", fill = T, header = T)

pheno_2743 <- read.table(gzfile("gtex/data/raw/phenotypes/phs000424.v9.pht002743.v9.p2.c1.GTEx_Sample_Attributes.GRU.txt.gz"),sep="\t", fill = T, header = T)
pheno_2743 <- pheno_2743[grepl("GTEX", pheno_2743$SAMPID), ]
subs <- lapply(strsplit(pheno_2743$SAMPID, "-"),"[[", 2)|>
                   unlist()

subs <- paste0("GTEX-",  subs)

pheno_2743$subs <- subs
pheno_2743 <- pheno_2743[which(pheno_2743$SAMPID %in% colnames(bulk)),] |>
                subset(SMMTRLTP == "RNA:Total RNA" & SMTSD == "Heart - Left Ventricle")



bulk.matches <- bulk[,colnames(bulk)[which(colnames(bulk) %in% pheno_2743$SAMPID)]]



#table(colnames(bulk) %in% pheno_2743$SAMPID)

#table(colnames(bulk) %in% pheno$SUBJID)
#pheno_2743[!pheno_2743$subs %in% pheno$SUBJID,]

# Select ID and basic pheno columns

basic.phenotypes <- c("SUBJID", "COHORT", "SEX", "AGE", "RACE", "ETHNCTY", "HGHT", "WGHT", "BMI")
basic.cols <- which(colnames(pheno) %in% basic.phenotypes)


# Select for columns indicating heart disease:
## MHHRTATT:  Heart attack, acute myocardial infarction, acute coronary syndrome
## MHHRTDIS:  Ischemic Heart Disease (coronary artery disease (CAD), coronary heart disease, ischemic cardiomyopathy)
## MHHRTDISB: Heart Disease
## MHHTN:     Hypertension
### Table of phenotypes here: ncbi.nlm.nih.gov/projects/gap/cgi-bin/dataset.cgi?study_id=phs000424.v7.p2&pht=2742

heart.disease <- c("MHHRTATT", 'MHHRTDIS', 'MHHRTDISB', 'MHHTN')
hd.cols <- which(colnames(pheno) %in% heart_disease )


# Subset to fewer phenotypes & subjects in our bulk
# heart_disease col specifies if a subject is positive for any hd phenotype. Subjects negative on all sub hd phenotypes will be negative total. Subjects missing all sub hd will be omitted for heart_disease.

pheno.basic <- pheno |>
               select(c(basic.cols, hd.cols)) |>
               subset(SUBJID %in% props.bisque$Sub) |>
               mutate(heart_disease = case_when(MHHRTATT == 1 | MHHRTDIS == 1 | MHHRTDISB == 1 | MHHTN == 1 ~ 1,
                                                MHHRTATT == 0 & MHHRTDIS == 0 & MHHRTDISB == 0 & MHHTN == 0 ~ 0))
                        

# subset to subjects in our bulk data

#write.csv(pheno.basic, "gtex/data/processed/internal/phenotypes/pheno_basic_lv")

```



## Using DirichletReg to model compositional changes



```{r dirichlet, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
library("DirichletReg")
# prep DirichletReg matrix
#https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/dataset.cgi?study_id=phs000424.v7.p2&pht=2742
dir.merged <- props.bisque |>
  merge(pheno.basic, by.x = "Sub", by.y = "SUBJID")

dir.merged["Prop"] <- dir.merged["Prop"] + 0.0001

#  Heart disease, Ischemic heart disease, hypertensive, or heart attack

dir.mat <- dir.merged |>
  dcast(Sub + COHORT + SEX + AGE + BMI + heart_disease ~ CellType, value.var = "Prop")

# Convert data to DirichletRegData object
dir.mat$CellTypes <- DR_data(dir.mat[,c(7:length(dir.mat))])

dir.mat$Sub <- as.factor(dir.mat$Sub)
dir.mat$COHORT <- as.factor(dir.mat$COHORT)
dir.mat$SEX <- as.factor(dir.mat$SEX)
dir.mat$AGE <- as.numeric(dir.mat$AGE)
dir.mat$BMI <- as.numeric(dir.mat$BMI)
dir.mat$heart_disease <- as.factor(dir.mat$heart_disease)

# Run Dirichlet regression
model.1 <- DirichReg(CellTypes ~ COHORT + SEX + AGE + BMI + heart_disease, data = dir.mat)

model.2 <- DirichReg(CellTypes ~ COHORT + SEX + AGE + BMI, data = dir.mat)

# compare models, find if interaction term improves model
anova(model.1, model.2)


# Pr(>Chi) is highly significant (p =  0.005655), which means there's strong evidence against the null hypothesis (the simple model is better). model.1 provides a significantly better fit to the data than model.2.
# this implies that the effect of the treatment may be different for different genotypes.
summary(model.1)
```



### Outro Notes


```{r music plot controls, fig.width = 12, fig.height= 6, echo=FALSE, message=FALSE, warning=FALSE}

library(forcats)
# Find the CellType with the largest average Prop
avg_prop <- decon.melt %>%
  group_by(CellType) %>%
  summarise(avg_prop = mean(Prop, na.rm = TRUE))

ref.cell <- avg_prop$CellType[[which.max(avg_prop$avg_prop)]]

decon.ref <- decon.melt |>
  subset(CellType == ref.cell)

rank <- decon.ref$Sub[order(decon.ref$Prop)]

# plot
ggplot(decon.melt,
       aes(x=factor(Sub), y=Prop, fill=CellType))  +
  geom_bar(stat='identity', 
           position = "fill", 
           width = 1) +
  scale_fill_brewer(name = "Cell Type",
                    palette = "Dark2")+
  ylab("Proportion") +
  facet_wrap(~version, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE))+ 
  theme(
    axis.text.x = element_blank(),
    strip.text = element_blank(),
    title = element_text(size = 20),
    legend.text = element_text(size = 18),
    axis.ticks.x = element_blank()) +
  xlab("GTEx Individuals")


geom_violin(
  mapping = NULL,
  data = NULL,
  stat = "ydensity",
  position = "dodge")

p <- decon.melt |>
ggplot(aes(x = CellType, y = Prop, fill = version)) +
  geom_boxplot(aes(fill = version), 
               width = 0.5, color = "black", outlier.shape = NA) +
  geom_point(position = position_jitterdodge(jitter.width = 0.05, jitter.height = 0, dodge.width = 0.5), size = 2, alpha = 0.5) +
  theme(axis.text.x = element_text(color = "black", size = 15, angle =15, vjust = 0.6), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks.x = element_line(color = "black", size = 1), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  ggtitle("MuSiC and Bisque gravitate towards different CM subtypes in GTEx")

p

```

```{r plot gtex COHORT, fig.width = 18, fig.height= 25, echo=FALSE, message=FALSE, warning=FALSE}

dir.merged$CellType <- reorder(dir.merged$CellType, -dir.merged$Prop, FUN = sum)



plots <- c()
plots[[1]] <- dir.merged |>
ggplot(aes(x = CellType, y = Prop, fill = COHORT)) +
  geom_boxplot(aes(fill = COHORT), 
               width = 0.5, color = "black", outlier.shape = NA) +
  geom_point(position = position_jitterdodge(jitter.width = 0.05, jitter.height = 0, dodge.width = 0.5), size = 2, alpha = 0.5) +
  theme(axis.text.x = element_text(color = "black", size = 15, angle =15, vjust = 0.6), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks.x = element_line(color = "black", size = 1), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  ggtitle("Heart composition in GTEx stratifies by post-mortem status")

# BMI plot
dir.merged$BMI <- as.numeric(dir.merged$BMI)
BMI_med <- median(as.numeric(dir.merged$BMI))
dir.merged$BMI_cat <- "upper" # make BMI upper and lower ranges
dir.merged$BMI_cat[dir.merged$BMI < BMI_med] <- "lower"

plots[[2]] <- dir.merged |>
ggplot(aes(x = CellType, y = Prop, fill = BMI_cat)) +
  geom_boxplot(aes(fill = BMI_cat), 
               width = 0.5, color = "black", outlier.shape = NA) +
  geom_point(position = position_jitterdodge(jitter.width = 0.05, jitter.height = 0, dodge.width = 0.5), size = 2, alpha = 0.5) +
  facet_wrap(~COHORT, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
  theme(axis.text.x = element_text(color = "black", size = 15, angle =15, vjust = 0.6), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks.x = element_line(color = "black", size = 1), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  ggtitle("BMI (above or below median)")

# Hypertensive plot

plots[[3]] <- dir.merged |>
ggplot(aes(x = CellType, y = Prop, fill = MHHTN)) +
  geom_boxplot(aes(fill = MHHTN), 
               width = 0.5, color = "black", outlier.shape = NA) +
  facet_wrap(~COHORT, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.05, jitter.height = 0, dodge.width = 0.5), size = 2, alpha = 0.5) +
  theme(axis.text.x = element_text(color = "black", size = 15, angle =15, vjust = 0.6), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks.x = element_line(color = "black", size = 1), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  ggtitle("Hypertensive status")


#  Heart disease, Ischemic heart disease, hypertensive, or heart attack
dir.merged <- dir.merged %>%
  rowwise() %>%
  mutate(heart_disease = as.factor(as.integer(any(c_across(c("MHHRTDISB", "MHHRTATT", "MHHRTDIS", "MHHTN")) == 1)))) %>%
  ungroup()


plots[[4]] <- dir.merged |>
ggplot(aes(x = CellType, y = Prop, fill = heart_disease)) +
  geom_boxplot(aes(fill = heart_disease), 
               width = 0.5, color = "black", outlier.shape = NA) +
  facet_wrap(~COHORT, 
             scales = "free_x",
             labeller =  label_wrap_gen(multi_line=FALSE)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.05, jitter.height = 0, dodge.width = 0.5), size = 2, alpha = 0.5) +
  theme(axis.text.x = element_text(color = "black", size = 15, angle =15, vjust = 0.6), # Increase size here for larger axis text
        axis.text.y = element_text(color = "black", size = 10), # Add this line for y-axis text size
        axis.ticks.x = element_line(color = "black", size = 1), # Add this line for thick black axis bars
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6), # Move legend to upper right
        panel.background = element_blank(), # Remove background gray color
        strip.text = element_text(size = 20),
        title = element_text(size = 20),
        legend.text = element_text(size = 18)) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype") +
  scale_fill_brewer(palette = "Dark2") + 
  ggtitle("Heart disease, ischemic heart disease, hypertensive, or heart attack")



wrap_plots(plots, ncol = 1)
```