---
title: 'BCVS: Jensen Collaboration'
author: "Brian Gural"
date: "7/19/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown



```{r load libs, message=FALSE, echo = T, warning=FALSE, results= 'hide', cache=FALSE, include=T}
# List libraries
libs <- c("Seurat", "ggplot2", "patchwork", "SeuratDisk","reshape2", "tidyverse",
          "SCpubr","shiny", "ggrepel", "gridExtra", "scCustomize", "httr", 
          "scales", "dplyr", "DeconvoBuddies", "readr", "SingleCellExperiment",
          "SummarizedExperiment", "Biobase", "ggmagnify", "stringr", "MuSiC",
          "ggforce", "DirichletReg", "readxl") # list libraries here
# Require all of them
lapply(libs, require, character.only = T)

# Also load in-house functions
source("jensen/scripts/functions/decon_all.R")
```


```{r load data, message=FALSE, echo = T, warning=FALSE, cache=FALSE, include=T}
# Script used to produce this can be found at jensen/scripts/rmd/by_date/07192023_1.rmd
sn <- LoadH5Seurat("jensen/data/processed/single_cell/merged_datasets/07192023/07192023_1.h5seurat")

# Load bulk data
bulk <- read_csv("jensen/data/processed/jensen_rau_froese_cpm") |>
  column_to_rownames(var = "...1")

# Load and format bulk data
jensen_bulk <- readxl::read_xlsx("jensen/data/raw/jensen_counts_correct.xlsx")[,-c(20,21)] |>
  as.data.frame()
colnames(jensen_bulk)[c(20,21)] <- c("WT MI", "KO MI")


```


```{r DeconvoBuddies Find markers, fig.width = 7, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# Update sn with annotations
cells <- sample(colnames(sn), 10000, replace = FALSE)
sn$ident <- Idents(sn)
seurat <- sn 
seurat$annots <- Idents(seurat)

# Subset Seurat object and convert to SummarizedExperiment
seurat <- seurat |>
  subset(cells = cells, features = rownames(seurat)[rownames(seurat) %in% rownames(bulk)])

sn.sce <- seurat |>
  as.SingleCellExperiment(assay = "RNA") |>
  as("SummarizedExperiment")

# Add gene names
rownames(sn.sce) <- rownames(seurat@assays$RNA@counts)

# Get markers based on expression ratios between clusters and 1vAll comparisons
ratios <- get_mean_ratio2(sn.sce, cellType_col = "ident")
fc <- findMarkers_1vAll(sn.sce, cellType_col = "ident", mod = "~origin")

# Subset markers based on rank_ratio and logFC
marker_genes <- ratios  |> 
  left_join(fc, by = c("gene", "cellType.target")) |>
  filter(gene %in% rownames(sn)) |>
  group_by(cellType.target) |>
  mutate(rank_ratio = row_number()) |>
  filter(rank_ratio <= 50) |> 
  arrange(desc(logFC), .by_group = TRUE) |>
  slice_head(n = 15) |>
  pull(gene)

rm(seurat)
gc()
```

```{r deconvolute music , echo=FALSE, warning=FALSE, include=T}
# Prep bulk matrix as ExpressionSet
bulk.es.exp <- bulk |> 
  as.matrix() %>%
  ExpressionSet(assayData = .) |> 
  exprs()

# Run MuSiC with subset genes
decon.melt <- EstimateCellTypeProportions(sn, bulk.es.exp, for.aitchison = FALSE, bulk.log2 = TRUE,
                                          sn.individuals = "orig.ident", marker = marker_genes, cells_exclude = NULL)
 
```

```{r prep phenos , fig.width = 14, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
# Load and format bulk data
jensen_bulk <- readxl::read_xlsx("jensen/data/raw/jensen_counts_correct.xlsx")[,-c(20,21)] |>
  as.data.frame()
colnames(jensen_bulk)[c(20,21)] <- c("WT MI", "KO MI")

phenotypes_real <- jensen_bulk[1,] |> 
  pivot_longer(cols = -c(1:5), 
               names_to = "gene_treat", 
               values_to = "id") |>
  mutate(gene_treat = str_remove(gene_treat, "\\..*")) |>
  select(gene_treat, id)

```

                              
```{r prep phenos , fig.width = 14, fig.height= 8, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(stringr)

# Turn music output into graph-friendly dataframe
decon.melt <- decon.melt |>
  mutate(
    CellType = factor(CellType, levels = unique(CellType)),
    Sub = factor(Sub, levels = unique(Sub)),
    Prop = as.numeric(Prop)
  )

# Bring in real phenotype data
decon.melt <- merge(decon.melt, phenotypes_real, by.x = "Sub", by.y = "id", all.x = TRUE) |>
  mutate(
    gene_treat = factor(gene_treat, levels = unique(gene_treat))
  )

# Add genotype info
decon.melt <- decon.melt |>
  mutate(
    genotype = factor(ifelse(!is.na(gene_treat), str_split(gene_treat, " ", simplify = TRUE)[,1], NA)),
    treatment = factor(ifelse(!is.na(gene_treat), str_split(gene_treat, " ", simplify = TRUE)[,2], NA)),
    type = factor(ifelse(str_detect(Sub, "RNA|Sham"), "fraction", "whole")),
    origin = factor(case_when(
      str_detect(Sub, "B6") ~ "Rau",
      str_detect(Sub, "Sham") ~ "Froese",
      TRUE ~ "Jensen"
    )),
    fraction_type = factor(case_when(
      str_detect(Sub, "CM") ~ "Cardiomyocytes",
      str_detect(Sub, "Fib|FB") ~ "Fibroblasts",
      str_detect(Sub, "Endo|EC_") ~ "Endothelial Cells",
      TRUE ~ NA_character_
    ))
  )



write.csv(decon.melt, "jensen/data/processed/outputs/compositions/bcsv_poster_ratios.csv", row.names = F)

```


```{r music plot sham, fig.width = 19, fig.height= 9, echo=FALSE, message=FALSE, warning=FALSE}

# Reorder the factor levels of CellType so that they are ordered by sum of their values accross groups
decon.melt$CellType <- reorder(decon.melt$CellType, -decon.melt$Prop, FUN = sum)

# Setting the levels
decon.melt$gene_treat <- factor(decon.melt$gene_treat, 
                                        levels = c('WT sham', 'WT MI', 'KO sham', 'KO MI'))

#! These colors are hard to read when printed, should be lighter
#! Or the dots from geom_jitter need to be bigger/brighter

my_palette <- c('#78c679', '#006d2c', '#d95f0e', '#993404')

plot.jen.sample <- decon.melt |>
  subset(type == "whole") |>
  ggplot(aes(x = as.numeric(CellType), y = Prop, fill = gene_treat)) +
  geom_bar(stat = "summary", 
           position = position_dodge(0.9), 
           fun = mean,
           width = 0.9,
           color = "black") +
  theme(axis.text.x = element_text(color = "black", size = 25, angle =15),
        axis.text.y = element_text(color = "black", size = 20),
        axis.ticks = element_blank(),
        legend.position = c(0.95, 0.95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6),
        strip.text = element_text(size = 20),
        title = element_text(size = 30),
        legend.text = element_text(size = 25),
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')) +
  labs(x = "Cell Type", 
       y = "Proportion", 
       fill = "Genotype-Treatment") +
  scale_fill_manual(values = my_palette) +  # using custom palette
  geom_jitter(inherit.aes = T, 
              position = position_jitterdodge(0.01, 0, 0.9),
              size = 2) +
  scale_x_continuous(
    breaks = 1:length(levels(decon.melt$CellType)),
    label = levels(decon.melt$CellType)) +
  geom_magnify(from = c(2.55,3.45,0,0.2), to = c(1.7,4,0.33,0.8), 
                   shape = "rect", shadow = TRUE) 

plot.jen.sample
```

```{r bcvs.fractions, fig.width = 16, fig.height= 4, echo=FALSE, message=FALSE, warning=FALSE}

# Plot the samples which are purified, bulk sequenced cell types
plot.fractions <- decon.melt |>
  subset(type == "fraction") |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 0.8,
           color = "black")+
   scale_fill_manual(values = pal)  +
   facet_wrap(~fraction_type, 
           scales = "free_x",
           labeller =  label_wrap_gen(multi_line=FALSE)) +
   ylab("Proportion") +
   theme_minimal() +
   theme(axis.text.x = element_blank(),
        axis.text.y = element_text(color = "black", size = 20),
        axis.ticks = element_blank(),
        strip.text = element_text(size = 30),
        title = element_text(size = 30),
        legend.text = element_text(size = 25),
        panel.background = element_blank(),
        panel.grid = element_blank()) +
   xlab("Sample replicates") 

plot.fractions

```



```{r dirichlet, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

## Prep DirichletReg matrix

# Add small value to remove zeros
decon.melt$Prop <- decon.melt$Prop + 0.0001

# Make model matrix
dir.mat <- decon.melt |>
  subset(type == "whole",
         select = c("Sub", "CellType", "Prop", "genotype", "treatment")) |>
  dcast(Sub + genotype + treatment ~ CellType, value.var = "Prop")


### TEMP ###

############

# Convert data to DirichletRegData object
dir.mat$CellTypes <- DR_data(dir.mat[,c(4:length(dir.mat))])

dir.mat$genotype   <- as.factor(dir.mat$genotype) |>
                        relevel(ref = "WT") 

dir.mat$treatment <- as.factor(dir.mat$treatment) |>
                        relevel(ref = "sham") 
# Run Dirichlet regression
model.1 <- DirichReg(CellTypes ~ treatment * genotype, data = dir.mat, model = "alternative", base = 5)
model.2 <- DirichReg(CellTypes ~ treatment * genotype, data = dir.mat, model = "common")

# compare models, find if interaction term improves model

summary(model.1)
model.1$sub.comp

```


```{r sig bars, fig.width = 8.7, fig.height= 10.5, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)

# !needs the pr and z value columns!
df <- data.frame(CellType = rep(c("Cardimyocytes", "Fibroblasts", 
                                  "Endothelial Cells", "Immune cells"),
                                times =1, each = 4),
                 Variable = rep(c("Intercept", "treatmentMI", 
                                  "genotypeKO", "treatmentMI:genotypeKO"),
                                times = 4),
                 coefficients = model.1$coefficients[-length(model.1$se)],
                 std_err = model.1$se[-length(model.1$se)])
# Create a data frame with the required information
df <- data.frame(
    CellType = c(
        "cardiomyocytes", "cardiomyocytes", "cardiomyocytes",
        "fibroblasts", "fibroblasts", "fibroblasts",
        "immune", "immune", "immune"
    ),
    Variable = c(
        "Myocardial Infarction", "a1-AR KO", "Myocardial Infarction and a1-AR KO",
        "Myocardial Infarction", "a1-AR KO", "Myocardial Infarction and a1-AR KO",
        "Myocardial Infarction", "a1-AR KO", "Myocardial Infarction and a1-AR KO"
    ),
    Estimate = c(0.27344, -0.02141, 0.01670, 1.0866, -0.0380, 0.5711, -1.1859, -0.1314, 1.6480),
    StdError = c(0.11116, 0.10321, 0.15849, 0.1658, 0.1791, 0.2313, 0.5499, 0.4366, 0.7057),
    zvalue = c(2.460, -0.207, 0.105, 6.555, -0.212, 2.469, -2.157, -0.301, 2.335),
    Pr = c(0.0139, 0.8357, 0.9161, 5.55e-11, 0.8320, 0.0135, 0.0310, 0.7635, 0.0195)
)

# Print the data frame
print(df)



# Add a column for significance labels
df <- df |>
    mutate(Significance = case_when(
        as.numeric(Pr ) < .001 ~ "***",
        as.numeric(Pr) < .01 ~ "**",
        as.numeric(Pr) < .05 ~ "*",
        as.numeric(Pr) < .1 ~ ".",
        TRUE ~ " "
    ))

# Load the required library
library(stringr)

# Wrap the Variable names
df$Variable <- str_wrap(df$Variable, width = 12)

# Now, plot the data
err.plot <- ggplot(df, aes(y = Variable, x = Estimate, color = CellType)) +
    geom_point(position = position_dodge(width = 0.6), size = 8) +
    geom_errorbarh(aes(xmin = Estimate - StdError, xmax = Estimate + StdError),
                  height = 0.2, position = position_dodge(width = 0.6), size = 2) +
    geom_text(aes(label = Significance, x = Estimate + StdError, group = CellType), 
              position = position_dodge(width = 0.6), 
              colour = "black", hjust = -0.2, size = 10) +
    scale_color_manual(values = pal) +
    theme_classic() +
    theme(axis.text.x = element_text(color = "black", size = 20),
        axis.text.y = element_text(color = "black", size = 20),
        axis.ticks = element_blank(),
        legend.position = c(1, 0.3),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(6, 6, 6, 6),
        title = element_text(size = 20),
        legend.text = element_text(size = 20),
        axis.title.y = element_text(color = "black", size = 20),
        axis.title.x = element_text(color = "black", size = 20),
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')) +
  labs(x = "Coeffiecients (estimates of effect)", 
       y = "Predictor Variables", 
       legend = "Cell Types") 

err.plot

```

```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
library(makeunique)
jensen_bulk <- read_xlsx("jensen/data/raw/jensen_counts_correct.xlsx")[-1,-c(1:4,20,21)] |>
               as.data.frame()
colnames(jensen_bulk) <- c("gene", make_unique(unlist(phenotypes_real$gene_treat)))

phenotypes_real$de_id <- colnames(jensen_bulk)[-1]

# Add a suffix to duplicate gene names to make them unique
make.names.jensen_bulk <- function(jensen_bulk){
    duplicated.names <- duplicated(jensen_bulk$gene)
    suffix <- cumsum(duplicated.names)
    jensen_bulk$gene[duplicated.names] <- paste0(jensen_bulk$gene[duplicated.names], ".", suffix[duplicated.names])
    return(jensen_bulk)
}

jensen_bulk <- make.names.jensen_bulk(jensen_bulk)

# Now set the row names
row.names(jensen_bulk) <- jensen_bulk$gene

write.csv(jensen_bulk, "jensen/results/deseq2_08032023/bulk_data")

write.csv(dir.mat$CellTypes, "jensen/results/deseq2_08032023/comp_data")

```

```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

# Prepare for DESeq2
count_data <- jensen_bulk[,-1]

count_data <- mutate_all(count_data, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers

# Prepare sample information
sample_info <- data.frame(
  row.names = colnames(count_data),
  genotype =  lapply(strsplit(colnames(count_data), " "),  "[[", 1) |> unlist() |> as.factor() ,
  treatment = lapply(strsplit(colnames(count_data), " "),  "[[", 2) |> unlist() |> as.factor()
)

sample_info$genotype <- relevel(sample_info$genotype, ref = "WT")
sample_info$treatment <- relevel(sample_info$treatment, ref = "sham")

# Make sure the sample information matches the count data
all(rownames(sample_info) == colnames(count_data))

# Create a DESeqDataSet
dds <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample_info,
  design = ~ genotype + treatment + genotype:treatment  # include interaction term
)

# Run the DESeq pipeline
dds <- DESeq(dds)

resultsNames(dds)

# Run the differential expression analysis
res <- results(dds, name="genotypeKO.treatmentMI")

```

```{r deseq w comp, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
sample_info_comp <- cbind(sample_info, dir.mat[,c(4:8)])
colnames(sample_info_comp)[5] <- "endothelial_cells"
# Make sure the sample information matches the count data
all(rownames(sample_info_comp) == colnames(count_data))

# Create DESeqDataSet w/ variable compositions in design
dds_composition  <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample_info_comp,
  design = ~ genotype + treatment + genotype:treatment + cardiomyocytes + fibroblasts + immune 
)

# Run 
dds_composition <- DESeq(dds_composition)

# Get interaction results
res_composition <- results(dds_composition, name = "genotypeKO.treatmentMI")

```

```{r compare deseq outs}
# Compare results
comparison <- merge(as.data.frame(res), as.data.frame(res_composition), by = "row.names", suffixes = c(".no_comp", ".comp"))

# Add a column indicating whether a gene is significant in either analysis
comparison$significant <- (comparison$pvalue.no_comp < 0.05) | (comparison$pvalue.comp < 0.05)

# Load ggplot2
library(ggplot2)

# Scatterplot of p-values
ggplot(comparison, aes(x = -log10(pvalue.no_comp), y = -log10(pvalue.comp), color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "-log10(p-value) without composition", y = "-log10(p-value) with composition")

# Scatterplot of log fold changes
ggplot(comparison, aes(x = log2FoldChange.no_comp, y = log2FoldChange.comp, color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "log2(Fold Change) without composition", y = "log2(Fold Change) with composition")

# Load necessary library
library(viridis)

# Calculate difference in -log10 p-values
comparison$p_diff <- -log10(comparison$pvalue.comp) - -log10(comparison$pvalue.no_comp)

# Generate volcano plot
ggplot(comparison, aes(x = log2FoldChange.comp, y = -log10(padj.comp), color = p_diff)) +
  geom_point(alpha = 0.5) +
  scale_color_viridis() +
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Difference in -log10(p-value)")+
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +  # add a line for p=0.05
  theme(legend.position = "bottom")  # move legend to the bottom


```

```{r deseq w comp graphs, fig.width = 11.5, fig.height= 10.7, echo=TRUE, message=TRUE, warning=FALSE}
# Load ggplot2
library(ggplot2)

# Scatterplot of p-values
ggplot(comparison, aes(x = -log10(pvalue.no_comp), y = -log10(pvalue.comp), color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "-log10(p-value) without composition", y = "-log10(p-value) with composition")

# Scatterplot of log fold changes
ggplot(comparison, aes(x = log2FoldChange.no_comp, y = log2FoldChange.comp, color = significant)) +
  geom_point(alpha = 0.5) +
  theme_minimal() +
  labs(x = "log2(Fold Change) without composition", y = "log2(Fold Change) with composition")

# Load necessary library
library(viridis)

# Calculate difference in -log10 p-values
comparison$p_diff <- -log10(comparison$pvalue.comp) - -log10(comparison$pvalue.no_comp)
hist(comparison$p_diff)


comparison <- comparison|>
                mutate(up_down = ifelse(p_diff >= 0,"up","down")) 
# Load necessary library
library(ggrepel)

# Get the gene names for the top 5 genes by padj.comp
top_genes <- comparison |> 
  group_by(up_down) |>  # take the top 5 rows
  arrange(desc(abs(p_diff)), .by_group = TRUE) |>
  na.omit() |>
  slice_head(n = 5) 

ggplot(comparison, aes(x = log2FoldChange.comp, y = -log10(padj.comp), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_viridis(option = "plasma") +
  geom_text_repel(data = top_genes, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) +
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank())  # move legend to the bottom

 # most changed genes

# Sort the dataframe by p_diff in ascending order
sorted_comparison <- comparison[order(comparison$p_diff), ]

# Display the top n genes that had the largest increase in p-value
n <- 10
top_n_genes <- head(sorted_comparison, n)
print(top_n_genes)

do_DimPlot(sn, reduction = "umap", label = T, repel = T)  +  
  NoLegend() + theme(title = element_text(size = 22))

```

```{r deseq w comp graphs, fig.width = 30, fig.height= 8, echo=TRUE, message=TRUE, warning=FALSE}
plotGene <- function(seurat, slot = "scale.data", gene, plot.title = T, plot.legend = T){
  
  seurat |>
    do_FeaturePlot(gene, reduction = "umap", slot = slot, 
                   legend.position = if(plot.legend == T){"bottom"}else{"none"}, 
                   legend.width = 2, legend.length = 25,  order = T) +
    theme(title = element_text(size = 18),
          legend.title = element_blank()) +
    labs(title = if(plot.title == T){paste0(gene, "\n-log10(p-value) change: " , names(gene))}else{" "})
}

genes_up <- top_genes |>
            subset(up_down == "up") |>
            pull(Row.names)

names(genes_up) <- top_genes |>
            subset(up_down == "up") |>
            pull(p_diff) |>
            round(digits = 2)

genes_down <- top_genes |>
            subset(up_down == "down") |>
            pull(Row.names)

names(genes_down) <- top_genes |>
            subset(up_down == "down") |>
            pull(p_diff) |>
            round(digits = 2)
# plot new
plots <- c()

plots[["cell"]] <- do_DimPlot(sn, reduction = "umap", label = T, repel = T)  +  
  NoLegend() + theme(title = element_text(size = 22))

plots.up <- lapply(names(genes_up), function(x){plotGene(seurat = sn, slot = "data", gene = genes_up[x], plot.legend = F)})

plots.down <- lapply(names(genes_down), function(x){plotGene(seurat = sn, slot = "data", gene = genes_down[x], plot.legend = F)})


plots.up <- c(plots, plots.up)

plots.down <- c(plots, plots.down)

wrap_plots(plots.up, 
           ncol = 6)
wrap_plots(plots.down, 
           ncol = 6)
```

