---
title: "DESeq Composition PCA vs CLR"
author: "Brian Gural"
date: "2023-08-05"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libs, message=FALSE, echo = F, warning=FALSE, results= 'hide', cache=FALSE, include=T}
# List libraries
libs <- c("Seurat", "ggplot2", "patchwork", "SeuratDisk","reshape2", "tidyverse",
          "SCpubr","shiny", "ggrepel", "gridExtra", "scCustomize", "httr", 
          "scales", "dplyr", "DeconvoBuddies", "readr", "SingleCellExperiment",
          "SummarizedExperiment", "Biobase", "ggmagnify", "stringr", "MuSiC",
          "ggforce", "DirichletReg", "readxl", "viridis", "DESeq2", "compositions") # list libraries here
# Require all of them
lapply(libs, require, character.only = T)

# Also load in-house functions
source("jensen/scripts/functions/decon_all.R")
```

```{r load data, cache = TRUE, echo=FALSE, message=F}
jensen_bulk <- read.csv("jensen/results/deseq2_08032023/bulk_data", row.names = 1)

comps <- read.csv("jensen/results/deseq2_08032023/comp_data", row.names = 1)
colnames(comps)[3] <- "endothelial_cells"
# Script used to produce this can be found at jensen/scripts/rmd/by_date/07192023_1.rmd
sn <- LoadH5Seurat("jensen/data/processed/single_cell/merged_datasets/07192023/07192023_1.h5seurat")
```

```{r clr transform}
# use clr or ilr to incorp the compositions into the design matrix
comps.clr <- clr(comps)

colnames(comps.clr) <- paste0("clr.", colnames(comps.clr))

```

```{r pca}

pca <- prcomp(t(comps[,-5]))

library(plotly)

plot(pca$rotation)
```

```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

# Prepare for DESeq2
count_data <- jensen_bulk[,-1]

count_data <- mutate_all(count_data, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers

# Prepare sample information
sample_info <- data.frame(
  row.names = colnames(count_data),
  genotype =  lapply(strsplit(colnames(count_data), "[.]"),  "[[", 1) |> unlist() |> as.factor() ,
  treatment = lapply(strsplit(colnames(count_data), "[.]"),  "[[", 2) |> unlist() |> as.factor()
)

sample_info$genotype <- relevel(sample_info$genotype, ref = "WT")
sample_info$treatment <- relevel(sample_info$treatment, ref = "sham")

```

```{r deseq pca}

sample.pca <- cbind(sample_info, pca$rotation)

# Create a DESeqDataSet
dds.pca <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample.pca,
  design = ~ genotype + treatment + genotype:treatment + PC1 
)

# Run the DESeq pipeline
dds.pca <- DESeq(dds.pca)

# Run the differential expression analysis
res.pca <- results(dds.pca, name="genotypeKO.treatmentMI")

```

```{r deseq clr}
sample.clr <- cbind(sample_info, comps.clr)

# Create a DESeqDataSet
dds.clr <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample.clr,
  design = ~ genotype + treatment + genotype:treatment + clr.cardiomyocytes + clr.endothelial_cells 
)

# Run the DESeq pipeline
dds.clr <- DESeq(dds.clr)

# Run the differential expression analysis
res.clr <- results(dds.clr, name="genotypeKO.treatmentMI")

```

```{r deseq raw}
# Create a DESeqDataSet
dds.raw <- DESeqDataSetFromMatrix(
  countData = count_data,
  colData = sample_info,
  design = ~ genotype + treatment + genotype:treatment
)

# Run the DESeq pipeline
dds.raw <- DESeq(dds.raw)

# Run the differential expression analysis
res.raw <- results(dds.raw, name="genotypeKO.treatmentMI")
```


```{r compare deseq pca, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
# Compare results
comparison.pca <- merge(as.data.frame(res.pca), as.data.frame(res.raw), by = "row.names", suffixes = c(".pca", ".raw"))

# Calculate difference in -log10 p-values
comparison.pca$p_diff <- -log10(comparison.pca$padj.pca) - -log10(comparison.pca$padj.raw)

# Make categorical column
comparison.pca <- comparison.pca |>
        mutate(pca.sig = ifelse(padj.pca <= 0.05 & padj.raw >= 0.05, T, F),
               raw.sig = ifelse(padj.pca >= 0.05 & padj.raw <= 0.05, T, F),
               both.sig = ifelse(padj.pca <= 0.05 & padj.raw <= 0.05, T, F)) 

# Get the gene names for the top 5 genes by change in adjusted p value, within lost and gained sig groups
gained.pca <- comparison.pca |>  
  subset(pca.sig == T) |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

lost.pca <- comparison.pca |>  
  subset(raw.sig == T) |> 
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

top.pca <- comparison.pca |> 
  arrange(padj.pca) |>
  slice_head(n = 5)
```

```{r plot pca, fig.width = 14, fig.height= 12}
limit <- comparison.pca |> 
  na.omit() |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 1) |>
  pull(p_diff) * c(1,-1)

plot.pca <- ggplot(comparison.pca, aes(x = log2FoldChange.pca, y = -log10(padj.pca), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_distiller(type = "div", limit = limit)+
  geom_text_repel(data = gained.pca, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 40, segment.linetype = 2, segment.size = 0.6, color = "orange", force_pull = 0.01, min.segment.length = 0) + # label most increase sig genes 
  geom_text_repel(data = lost.pca, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "purple", force_pull = 0.01, min.segment.length = 0) + # label most decrease sig genes 
  geom_text_repel(data = top.pca, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) + # label most sig genes 
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
        panel.background = element_rect(fill='transparent'),
        plot.background = element_rect(fill='transparent', color=NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  ggtitle("DE w/ PCA-based composition in model", subtitle = "Colored text show genes that became or lost significance, top 5 by change")# move legend to the bottom

plot.pca
```

```{r compare deseq clr, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
# Compare results
comparison.clr <- merge(as.data.frame(res.clr), as.data.frame(res.raw), by = "row.names", suffixes = c(".clr", ".raw"))

# Calculate difference in -log10 p-values
comparison.clr$p_diff <- -log10(comparison.clr$padj.clr) - -log10(comparison.clr$padj.raw)

# Make categorical column
comparison.clr <- comparison.clr |>
        mutate(clr.sig = ifelse(padj.clr <= 0.05 & padj.raw >= 0.05, T, F),
               raw.sig = ifelse(padj.clr >= 0.05 & padj.raw <= 0.05, T, F),
               both.sig = ifelse(padj.clr <= 0.05 & padj.raw <= 0.05, T, F)) 

# Get the gene names for the top 5 genes by change in adjusted p value, within lost and gained sig groups
gained.clr <- comparison.clr |>  
  subset(clr.sig == T) |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

lost.clr <- comparison.clr |>  
  subset(raw.sig == T) |> 
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

top.clr <- comparison.clr |> 
  arrange(padj.clr) |>
  slice_head(n = 5)
```

```{r plot clr, fig.width = 14, fig.height= 12}
limit <- comparison.clr |> 
  na.omit() |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 1) |>
  pull(p_diff) * c(-1, 1)

plot.clr <- ggplot(comparison.clr, aes(x = log2FoldChange.clr, y = -log10(padj.clr), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_distiller(type = "div", limit = limit)+
  geom_text_repel(data = gained.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 40, segment.linetype = 2, segment.size = 0.6, color = "orange", force_pull = 0.01, min.segment.length = 0) + # label most increase sig genes 
  geom_text_repel(data = lost.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "purple", force_pull = 0.01, min.segment.length = 0) + # label most decrease sig genes 
  geom_text_repel(data = top.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) + # label most sig genes 
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
        panel.background = element_rect(fill='transparent'),
        plot.background = element_rect(fill='transparent', color=NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  ggtitle("DE w/ clr-based composition in model", subtitle = "Colored text show genes that became or lost significance, top 5 by change")# move legend to the bottom

plot.clr
```

